---
output: html_document
params:
  cpus: 1
  artifact_dir: "/data/scratch/sturm/tmp"
  meta: NULL
  scoring_function: "mean_z"
---

```{r}
library(conflicted)
library(dplyr)
conflict_prefer("filter", "dplyr")
library(readxl)
library(survminer)
library(survival)
library(tidyr)
library(readr)
library(stringr)
# library(immunedeconv)
library(tibble)
library(omnideconv)
library(ggbeeswarm)
library(cowplot)
```

# Load data

```{r, include=FALSE}
clinical_data = read_tsv("../../tables/tcga/clinical_data_for_scissor.tsv")
tcga_tpm = read_rds("../../data/13_tcga/for_scissor/nsclc_primary_tumor.rds")
age = read_excel("../../tables/tcga/mmc1.xlsx") %>% select(TCGA_patient_barcode=bcr_patient_barcode, age=age_at_initial_pathologic_diagnosis)
tan_deseq2 = read_tsv("../../data/30_downstream_analyses/de_analysis/tumor_normal/de_deseq2/adata_tumor_normal_neutrophils_DESeq2_result.tsv")
```

```{r}
# `folder_cho`
load("../../data/14_ici_treatment/all_cho.RData")
# `folder_jung`
load("../../data/14_ici_treatment/all_jung.RData")

ici_meta = bind_rows(
  folder_cho$clinical_response |> 
    mutate(type=if_else(str_trim(Histology)=="Adenocarcinoma", "LUAD", "LUSC"), response=str_trim(Responsiveness)) |>
    select(type, age=Age, sex=Sex, response), 
  folder_jung$clinical_response |>
    mutate(response=if_else(str_trim(Clinical.benefit) == "NDB", "Non-responder", "Responder")) |>
    select(response)
)
ici_tpm = bind_cols(folder_cho$tpm, folder_jung$tpm)[, rownames(ici_meta)]
```

# Signature scoring

```{r}
# These are genes highly specific for the Neutrophil cluster in primary tumor samples
neutro_sig = c('ADGRG3',
 'BCL6',
 'CMTM2',
 'CSF3R',
 'CXCR1',
 'CXCR2',
 'FCGR3B',
 'FFAR2',
 'HCAR3',
 'IFITM2',
 'IL1R2',
 'IVNS1ABP',
 'LITAF',
 'MBD6',
 'MX2',
 'NAMPT',
 'NSMAF',
 'PROK2',
 'R3HDM4',
 'RNF149',
 'SLC25A37',
 'SMCHD1',
 'TRIB1',
 'TRIM25',
 'USP15')

# These genes are a subset of the above. They are fewer, but even more specific. 
neutro_sig2 = c('ADGRG3', 'CMTM2', 'CXCR1', 'CXCR2', 'FCGR3B', 'PROK2')

neutro_sig3 = c("CMTM", "PROK2", "ADGRG3")

# These are genes relatively specific for TANs vs. NANs and all other 
# cell-types based on single-cell data from primary tumor and normal adjacent
# data
tan_sig = c('ADGRG3',
 'BCL2A1',
 'CLEC4E',
 'DDX60L',
 'GK',
 'HCAR3',
 'IER3',
 'ITGAX',
 'LCP2',
 'LILRA2',
 'NFKBIZ',
 'NSMAF',
 'PHACTR1',
 'PPIF',
 'SLC11A1')

# These are the top genes differentially expressed between TANs and NANs
# They are not necessarily specific for the Neutrophil cluster. 
tan_sig2 = tan_deseq2 |> filter(log2FoldChange > 2) |> arrange(padj) |> head(20) |> pull(gene_id...2)

# LCAM signature from the Merad paper
lcam_tab = read_excel("../../tables/gene_annotations/merad_lcam.xlsx", sheet="LCAM axis", col_names = c("signature", "genes"))
lcam = str_split(lcam_tab$genes, ",")
names(lcam) = lcam_tab$signature
```

```{r}
#' Score signature genes as defined in the MCP counter paper
#'
#' This is a simple mean of the log transformed gene expression (aka geometric mean)
get_mcp_scores = function(mixture, genes) {
  exclude_genes = setdiff(genes, rownames(mixture))
  if(length(exclude_genes)) {
    warning(sprintf("Excluded the following genes which are not in the mixture matrix: %s", paste(exclude_genes, collapse=", ")))
    genes = setdiff(genes, exclude_genes)
  }
  signature_expression = log1p(mixture[genes, ])
  apply(signature_expression, MARGIN = 2, mean, na.rm=TRUE)
}

#' Score signature genes as described in the Merad paper
#' 
#' This is a simple mean of the z-scored gene expression
get_lcam_scores = function(mixture, genes) {
  exclude_genes = setdiff(genes, rownames(mixture))
  if(length(exclude_genes)) {
    warning(sprintf("Excluded the following genes which are not in the mixture matrix: %s", paste(exclude_genes, collapse=", ")))
    genes = setdiff(genes, exclude_genes)
  }
  signature_expression = t(scale(t(log1p(mixture[genes, ]))))
  apply(signature_expression, MARGIN = 2, mean, na.rm=TRUE)
}

SCORING_FUNCTIONS = list("mean"=get_mcp_scores, "mean_z"=get_lcam_scores)
scoring_function = SCORING_FUNCTIONS[[params$scoring_function]]
```

## LUAD vs LUSC

In TCGA: 

```{r, fig.width=10, fig.height=6}
tcga_scores = data.frame(
  sig_neutro=scoring_function(tcga_tpm, neutro_sig), 
  sig_neutro2=scoring_function(tcga_tpm, neutro_sig2),
  sig_neutro3=scoring_function(tcga_tpm, neutro_sig3),
  sig_tan=scoring_function(tcga_tpm, tan_sig),
  sig_tan2=scoring_function(tcga_tpm, tan_sig2),
  sig_lcam_lo=scoring_function(tcga_tpm, lcam$`LCAM-lo genes`),
  sig_lcam_hi=scoring_function(tcga_tpm, lcam$`LCAM-hi genes`)
) |> 
  mutate(sig_lcam_consensus = sig_lcam_hi - sig_lcam_lo) |>
  as_tibble(rownames="TCGA_patient_barcode")

tcga_clin_and_scores = clinical_data %>% 
  left_join(age) %>%
  inner_join(tcga_scores) 

tcga_clin_and_scores |> select(type, starts_with("sig_")) |>
  pivot_longer(-type, names_to="signature", values_to="score") |>
  ggplot(aes(x=type, y=score)) + 
  geom_quasirandom(aes(color=type)) + 
  geom_boxplot(outlier.alpha=0, width=.2) + 
  stat_compare_means(method="wilcox.test") +
  theme_cowplot() + 
  background_grid() + 
  facet_wrap(~signature, ncol = 4) + 
  ggtitle(sprintf("Scoring function: %s", params$scoring_function))
```

In response data:

```{r, fig.width=10, fig.height=6}
  ici_scores = data.frame(
    sig_neutro=scoring_function(ici_tpm, neutro_sig), 
    sig_neutro2=scoring_function(ici_tpm, neutro_sig2),
    sig_neutro3=scoring_function(ici_tpm, neutro_sig3),
    sig_tan=scoring_function(ici_tpm, tan_sig),
    sig_tan2=scoring_function(ici_tpm, tan_sig2),
    sig_lcam_lo=scoring_function(ici_tpm, lcam$`LCAM-lo genes`),
    sig_lcam_hi=scoring_function(ici_tpm, lcam$`LCAM-hi genes`)
  ) |> 
    mutate(sig_lcam_consensus = sig_lcam_hi - sig_lcam_lo) |>
    as_tibble(rownames="sample")
  
  ici_clin_and_scores = ici_meta |> 
    as_tibble(rownames="sample") |>
    inner_join(ici_scores) 
  
  ici_clin_and_scores |> select(type, starts_with("sig_")) |>
    filter(!is.na(type)) |> 
    pivot_longer(-type, names_to="signature", values_to="score") |>
    ggplot(aes(x=type, y=score)) + 
    geom_boxplot(outlier.alpha=0, width=.2) + 
    geom_quasirandom(aes(color=type)) + 
    stat_compare_means(method="wilcox.test") +
    theme_cowplot() + 
    background_grid() + 
    facet_wrap(~signature, ncol = 4) + 
    ggtitle(sprintf("Scoring function: %s", params$scoring_function))

```

## Survival analysis

```{r}
lapply(colnames(select(tcga_clin_and_scores, starts_with("sig"))), function(x) {
  coxph(as.formula(sprintf("Surv(time, status) ~ %s + type + age + tumor_stage", x)), data = tcga_clin_and_scores)
})
```

```{r, fig.width=6, fig.height=6}
plot_kaplan_meyer = function(clinical_data, column, q=c(0.25, .75)) {
  tmp_col = clinical_data[[column]]
  cutoffs = quantile(tmp_col, q)
  group = rep(NA, length(tmp_col))
  group[tmp_col < cutoffs[1]] = "low"
  group[tmp_col > cutoffs[2]] = "high"
  clinical_data$group = group
  clinical_data = clinical_data |> filter(time < 365 * 5)
  fit = survfit(Surv(time, status) ~ group, data=clinical_data)
  ggsurvplot(fit, data=clinical_data, pval=TRUE, risk.table = TRUE)
}

plot_kaplan_meyer(tcga_clin_and_scores |> filter(type == "LUAD"), "sig_neutro")
plot_kaplan_meyer(tcga_clin_and_scores |> filter(type == "LUSC"), "sig_neutro")

```

## Responder vs. Non-responder

```{r, fig.width=12, fig.height=6}
ici_clin_and_scores |> select(type, response, starts_with("sig_")) |>
    pivot_longer(-c(type, response), names_to="signature", values_to="score") |>
    ggplot(aes(x=response, y=score)) + 
    geom_boxplot(outlier.alpha=0, width=.2) + 
    geom_quasirandom(aes(color=type)) + 
    stat_compare_means(method="wilcox.test") +
    theme_cowplot() + 
    background_grid() + 
    facet_wrap(~signature, ncol = 4) + 
    ggtitle(sprintf("Scoring function: %s", params$scoring_function))
```

```{r, fig.width=5, fig.height=4}
ici_clin_and_scores |> 
  ggplot(aes(y=sig_lcam_hi, x=sig_neutro)) + geom_smooth(method="lm", color="black") + geom_point(aes(color=type)) + stat_cor() + theme_cowplot() + background_grid()

ici_clin_and_scores |> 
  ggplot(aes(y=sig_tan, x=sig_neutro)) + geom_smooth(method="lm", color="black") + geom_point(aes(color=type)) + stat_cor() + theme_cowplot() + background_grid()

ici_clin_and_scores |> 
  ggplot(aes(y=sig_lcam_hi, x=sig_tan)) + geom_smooth(method="lm", color="black") + geom_point(aes(color=type)) + stat_cor() + theme_cowplot() + background_grid()
```
