---
output: html_document
params:
  cpus: 1
  artifact_dir: "/data/scratch/sturm/tmp"
  meta: NULL
  scoring_function: "mean_z"
---

```{r}
library(conflicted)
library(dplyr)
conflict_prefer("filter", "dplyr")
library(readxl)
library(survminer)
library(survival)
library(tidyr)
library(readr)
library(stringr)
library(immunedeconv)
conflict_prefer("deconvolute", "immunedeconv")
library(tibble)
# library(omnideconv)
library(ggbeeswarm)
library(cowplot)
library(corrplot)
```

# Load data

```{r, include=FALSE}
clinical_data = read_tsv("../../tables/tcga/clinical_data_for_scissor.tsv")
tcga_tpm = read_rds("../../data/13_tcga/for_scissor/nsclc_primary_tumor.rds")
age = read_excel("../../tables/tcga/mmc1.xlsx") %>% select(TCGA_patient_barcode=bcr_patient_barcode, age=age_at_initial_pathologic_diagnosis)
tan_deseq2 = read_tsv("../../data/30_downstream_analyses/de_analysis/tumor_normal/de_deseq2/adata_tumor_normal_neutrophils_DESeq2_result.tsv")
tcga_treatment = read_tsv("../../tables/tcga/tcga_treatment.tsv")
```

```{r}
# `folder_cho`
load("../../data/14_ici_treatment/all_cho.RData")
# `folder_jung`
load("../../data/14_ici_treatment/all_jung.RData")

ici_meta = bind_rows(
  folder_cho$clinical_response |> 
    mutate(type=if_else(str_trim(Histology)=="Adenocarcinoma", "LUAD", "LUSC"), response=str_trim(Responsiveness)) |>
    select(type, age=Age, sex=Sex, response) |>
    mutate(dataset="cho"),
  folder_jung$clinical_response |>
    mutate(response=if_else(str_trim(Clinical.benefit) == "NDB", "Non-responder", "Responder")) |>
    select(response) |> 
    mutate(dataset="jung")
)
ici_tpm = bind_cols(folder_cho$tpm, folder_jung$tpm)[, rownames(ici_meta)]

clinical_data = clinical_data |>
  mutate(ajcc_stage = as.factor(recode(ajcc_pathologic_tumor_stage, `Stage IV` = 4, `Stage IIIA` = 3, `Stage IB` = 1, `Stage IA` = 1, `Stage I` = 1, `Stage II` = 2, `Stage IIA` = 2, `Stage IIB` = 2, `Stage III`= 3, `Stage IIIB` = 3, .default=NA_real_)))

selected_treatments = tcga_treatment |> 
  filter(tumor %in% c("LUAD", "LUSC"), !is.na(bcr_drug_barcode)) |>
  select(drug_name, measure_of_response, type=tumor, patient_barcode, therapy_type, treatment_n) |>
  filter(treatment_n == 1) |> # only 1st line 
  mutate(TCGA_patient_barcode = str_to_upper(patient_barcode)) |> 
  filter(therapy_type == "chemotherapy", !is.na(measure_of_response))
```

# Signature scoring

```{r}
# # These are genes highly specific for the Neutrophil cluster in primary tumor samples
# neutro_sig = c('ADGRG3',
#  'BCL6',
#  'CMTM2',
#  'CSF3R',
#  'CXCR1',
#  'CXCR2',
#  'FCGR3B',
#  'FFAR2',
#  'HCAR3',
#  'IFITM2',
#  'IL1R2',
#  'IVNS1ABP',
#  'LITAF',
#  'MBD6',
#  'MX2',
#  'NAMPT',
#  'NSMAF',
#  'PROK2',
#  'R3HDM4',
#  'RNF149',
#  'SLC25A37',
#  'SMCHD1',
#  'TRIB1',
#  'TRIM25',
#  'USP15')
# 
# # These genes are a subset of the above. They are fewer, but even more specific. 
# neutro_sig2 = c('ADGRG3', 'CMTM2', 'CXCR1', 'CXCR2', 'FCGR3B', 'PROK2')
# 
# neutro_sig3 = c("CMTM", "PROK2", "ADGRG3")
# neutro_sig3 = c("FCGR3B")
# 
sig_pd1 = c("PDCD1")
# 
# # These are genes relatively specific for TANs vs. NANs and all other 
# # cell-types based on single-cell data from primary tumor and normal adjacent
# # data
# tan_sig = c('ADGRG3',
#  'BCL2A1',
#  'CLEC4E',
#  'DDX60L',
#  'GK',
#  'HCAR3',
#  'IER3',
#  'ITGAX',
#  'LCP2',
#  'LILRA2',
#  'NFKBIZ',
#  'NSMAF',
#  'PHACTR1',
#  'PPIF',
#  'SLC11A1')
# 
# # These are the top genes differentially expressed between TANs and NANs
# # They are not necessarily specific for the Neutrophil cluster. 
# tan_sig2 = tan_deseq2 |> filter(log2FoldChange > 2) |> arrange(padj) |> head(20) |> pull(gene_id...2)

neutro_sig_tab = read_csv("../../data/30_downstream_analyses/neutrophils/neutro_sigs.csv")
neutro_sig = sapply(unique(neutro_sig_tab$signature), function(sig) { neutro_sig_tab |> filter(signature == !!sig) |> pull(gene_symbol)}, USE.NAMES = TRUE, simplify = FALSE)


# LCAM signature from the Merad paper
lcam_tab = read_excel("../../tables/gene_annotations/merad_lcam.xlsx", sheet="LCAM axis", col_names = c("signature", "genes"))
lcam = str_split(lcam_tab$genes, ",")
names(lcam) = lcam_tab$signature

pmn_mdsc = read_csv("../../tables/gene_annotations/pmn_mdsc.csv", comment = "#", col_names="gene_symbol")$gene_symbol
b_cell_tab = read_csv("../../tables/gene_annotations/b_cell_signatures.csv", comment="#")
b_cell = sapply(unique(b_cell_tab$signature), function(sig) { b_cell_tab |> filter(signature == !!sig) |> pull(gene_symbol)}, USE.NAMES = TRUE, simplify = FALSE)
```

```{r}
norm01 = function(x)  (x-min(x))/(max(x)-min(x))

#' Score signature genes as defined in the MCP counter paper
#'
#' This is a simple mean of the log transformed gene expression (aka geometric mean)
get_mcp_scores = function(mixture, genes) {
  exclude_genes = setdiff(genes, rownames(mixture))
  if(length(exclude_genes)) {
    warning(sprintf("Excluded the following genes which are not in the mixture matrix: %s", paste(exclude_genes, collapse=", ")))
    genes = setdiff(genes, exclude_genes)
  }
  signature_expression = log1p(mixture[genes, ])
  apply(signature_expression, MARGIN = 2, mean, na.rm=TRUE)
}

#' Score signature genes as described in the Merad paper
#' 
#' This is a simple mean of the z-scored gene expression
get_lcam_scores = function(mixture, genes) {
  exclude_genes = setdiff(genes, rownames(mixture))
  if(length(exclude_genes)) {
    warning(sprintf("Excluded the following genes which are not in the mixture matrix: %s", paste(exclude_genes, collapse=", ")))
    genes = setdiff(genes, exclude_genes)
  }
  signature_expression = t(scale(t(log1p(mixture[genes, ,drop=FALSE]))))
  apply(signature_expression, MARGIN = 2, mean, na.rm=TRUE)
}

SCORING_FUNCTIONS = list("mean"=get_mcp_scores, "mean_z"=get_lcam_scores)
scoring_function = SCORING_FUNCTIONS[[params$scoring_function]]
```

## LUAD vs LUSC

In TCGA: 

```{r, fig.width=13, fig.height=8}
tcga_deconv = deconvolute(tcga_tpm, method="mcp_counter") |> column_to_rownames("cell_type") |> as.matrix() |> t() |> scale() |> as_tibble(rownames="TCGA_patient_barcode") |> rename_at(vars(-TCGA_patient_barcode), function(x) paste0("ct_", x))

tcga_scores = data.frame(
  sig_neutro=scoring_function(tcga_tpm, neutro_sig$neutro_sig), 
  sig_neutro2=scoring_function(tcga_tpm, neutro_sig$neutro_sig2),
  sig_neutro3=scoring_function(tcga_tpm, neutro_sig$neutro_sig3),
  sig_tan=scoring_function(tcga_tpm, neutro_sig$tan_sig),
  sig_nan=scoring_function(tcga_tpm, neutro_sig$nan_sig),
  sig_lcam_lo=scoring_function(tcga_tpm, lcam$`LCAM-lo genes`),
  sig_lcam_hi=scoring_function(tcga_tpm, lcam$`LCAM-hi genes`),
  sig_pd1=scoring_function(tcga_tpm, sig_pd1),
  sig_pmn_mdsc=scoring_function(tcga_tpm, pmn_mdsc),
  sig_b_cell_follicular=scoring_function(tcga_tpm, b_cell$`Follicular B cells`),
  sig_b_cell_gc=scoring_function(tcga_tpm, b_cell$`Germinal center B cells`),
  sig_plasma=scoring_function(tcga_tpm, b_cell$`Plasma cells`)
) |> 
  mutate(sig_lcam_consensus = norm01(sig_lcam_hi) - norm01(sig_lcam_lo), 
         sig_tan_nan=(sig_tan + sig_nan) / 2) |>
  as_tibble(rownames="TCGA_patient_barcode")

tcga_clin_and_scores = clinical_data %>% 
  left_join(age) %>%
  inner_join(tcga_scores) %>%
  left_join(tcga_deconv) %>%
  left_join(selected_treatments)

tcga_clin_and_scores |> select(type, starts_with("sig_")) |>
  pivot_longer(-type, names_to="signature", values_to="score") |>
  ggplot(aes(x=type, y=score)) + 
  geom_quasirandom(aes(color=type)) + 
  geom_boxplot(outlier.alpha=0, width=.2) + 
  stat_compare_means(method="wilcox.test") +
  theme_cowplot() + 
  background_grid() + 
  facet_wrap(~signature, ncol = 5) + 
  ggtitle(sprintf("Scoring function: %s", params$scoring_function))
```
```{r, fig.width=13, fig.height=8}
tcga_clin_and_scores |> select(tumor_stage, starts_with("sig_")) |>
  mutate(tumor_stage = as.factor(tumor_stage)) |> 
  filter(!is.na(tumor_stage)) |> 
  pivot_longer(-tumor_stage, names_to="signature", values_to="score") |>
  ggplot(aes(x=tumor_stage, y=score)) + 
  geom_quasirandom(aes(color=tumor_stage)) + 
  geom_boxplot(outlier.alpha=0, width=.2) + 
  stat_compare_means(method="wilcox.test") +
  theme_cowplot() + 
  background_grid() + 
  facet_wrap(~signature, ncol = 5) + 
  ggtitle(sprintf("Scoring function: %s", params$scoring_function))

```

```{r, fig.width=10, fig.height=6}
tcga_clin_and_scores |> select(type, ajcc_stage, starts_with("sig_"), starts_with("ct_")) |>
  filter(!is.na(ajcc_stage)) |> 
  pivot_longer(-c(type, ajcc_stage), names_to="signature", values_to="score") |>
  filter(signature == "sig_neutro") |> 
  ggplot(aes(x=ajcc_stage, y=score)) + 
  geom_quasirandom(aes(color=ajcc_stage)) + 
  geom_boxplot(outlier.alpha=0, width=.2, alpha=.7) + 
  theme_cowplot() + 
  background_grid() + 
  ggtitle(sprintf("Scoring function: %s", params$scoring_function)) + 
  facet_wrap(~type) + 
  stat_compare_means(method="anova")
```

```{r, fig.width=6, fig.height=6}
sig_codes = Vectorize(function(p) {
  if(p < 0.001) {
    "p < 0.001"
  } else if(p < 0.01) {
    "p < 0.01"
  } else if (p < 0.05) {
    "p < 0.05"
  } else if (p < 0.1) {
    "p < 0.1"
  } else {
    "p >= 0.1"
  }
})

chemo_data = tcga_clin_and_scores |> 
  filter(measure_of_response %in% c("clinical progressive disease", "complete response"))  |>
  mutate(measure_of_response = as.factor(measure_of_response))

glm_res = lapply(colnames(select(chemo_data, starts_with("sig_"), starts_with("ct_"))), function(col) {
  res = summary(glm(as.formula(sprintf("measure_of_response ~ `%s` + type + ajcc_stage", col)), family="binomial", data=chemo_data))
  res$coefficients |> as_tibble(rownames="var") |> filter(grepl( col, var)) |> mutate(var=col)
}) |> bind_rows() |> mutate(sig_code = sig_codes(`Pr(>|z|)`)) |> arrange(-Estimate) |> mutate(var=factor(var, levels=unique(var)))

glm_res |> ggplot(aes(x=var, y=Estimate)) + geom_bar(aes(fill=sig_code), stat="identity") + scale_fill_brewer(palette= "Blues", direction=-1)  + theme_cowplot()+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + geom_hline(yintercept = 0, color="black")
```

```{r, fig.width=5, fig.height=7}
chemo_data |> select(measure_of_response, type, `ct_T cell CD8+`, sig_neutro) |> pivot_longer(-c(type, measure_of_response), names_to="signature", values_to="value") |> 
  ggplot(aes(x=measure_of_response, y=value)) + geom_boxplot(outlier.alpha=0, width=.2) + 
    geom_quasirandom(aes(color=type), size=0.5) + 
    stat_compare_means(method="wilcox.test") +
    theme_cowplot() + 
    background_grid() + 
    facet_wrap(~signature, ncol = 5) + 
    ggtitle(sprintf("Scoring function: %s", params$scoring_function)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

In response data:

```{r, fig.width=9, fig.height=6}

ici_deconv = deconvolute(ici_tpm, method="mcp_counter") |> column_to_rownames("cell_type") |> as.matrix() |> t() |> scale() |> as_tibble(rownames="sample") |> rename_at(vars(-sample), function(x) paste0("ct_", x))

ici_scores = data.frame(
    sig_neutro=scoring_function(ici_tpm, neutro_sig$neutro_sig), 
    sig_neutro2=scoring_function(ici_tpm, neutro_sig$neutro_sig2),
    sig_neutro3=scoring_function(ici_tpm, neutro_sig$neutro_sig3),
    sig_tan=scoring_function(ici_tpm, neutro_sig$tan_sig),
    sig_nan=scoring_function(ici_tpm, neutro_sig$nan_sig),
    sig_lcam_lo=scoring_function(ici_tpm, lcam$`LCAM-lo genes`),
    sig_lcam_hi=scoring_function(ici_tpm, lcam$`LCAM-hi genes`),
    sig_pd1=scoring_function(ici_tpm, sig_pd1),
    sig_pmn_mdsc=scoring_function(ici_tpm, pmn_mdsc),
    sig_b_cell_follicular=scoring_function(ici_tpm, b_cell$`Follicular B cells`),
    sig_b_cell_gc=scoring_function(ici_tpm, b_cell$`Germinal center B cells`),
    sig_plasma=scoring_function(ici_tpm, b_cell$`Plasma cells`)
) |> 
  mutate(sig_lcam_consensus = norm01(sig_lcam_hi) - norm01(sig_lcam_lo)) |>
  mutate(sig_combined = sig_lcam_hi + sig_neutro) |>
  mutate(sig_tan_nan=(sig_tan + sig_nan) / 2) |>
  as_tibble(rownames="sample")

ici_clin_and_scores = ici_meta |> 
  as_tibble(rownames="sample") |>
  inner_join(ici_scores) |>
  inner_join(ici_deconv) 

ici_clin_and_scores |> select(type, starts_with("sig_")) |>
  filter(!is.na(type)) |> 
  pivot_longer(-type, names_to="signature", values_to="score") |>
  ggplot(aes(x=type, y=score)) + 
  geom_boxplot(outlier.alpha=0, width=.2) + 
  geom_quasirandom(aes(color=type)) + 
  stat_compare_means(method="wilcox.test") +
  theme_cowplot() + 
  background_grid() + 
  facet_wrap(~signature, ncol = 5) + 
  ggtitle(sprintf("Scoring function: %s", params$scoring_function))

```

## Survival analysis

```{r, fig.width=6, fig.height=6}
lapply(c("all", "LUSC", "LUAD"), function(type) {
  filter_type = if(type == "all") {c("LUAD", "LUSC") } else {type}
  type_covar = ifelse(type == "all", "+ type ", "")
  coxph_res = lapply(colnames(select(chemo_data, starts_with("sig_"), starts_with("ct_"))), function(col) {
     res = summary(coxph(as.formula(sprintf("Surv(time, status) ~ `%s` %s + ajcc_stage", col, type_covar)), data=tcga_clin_and_scores |> filter(type %in% filter_type)))
     res$coefficients |> as_tibble(rownames="var") |> filter(grepl( col, var)) |> mutate(var=col, type=type)
  }) |> bind_rows() |> mutate(sig_code = sig_codes(`Pr(>|z|)`)) |> arrange(-coef) |> mutate(var=factor(var, levels=unique(var)))
  
  coxph_res |> ggplot(aes(x=var, y=coef)) + geom_bar(aes(fill=sig_code), stat="identity") + scale_fill_brewer(palette= "Blues", direction=-1)  + theme_cowplot()+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + geom_hline(yintercept = 0, color="black") + facet_wrap(~type, ncol=1)
})
```

```{r, fig.width=6, fig.height=6}
plot_kaplan_meyer = function(clinical_data, column, q=c(0.25, .75)) {
  tmp_col = clinical_data[[column]]
  cutoffs = quantile(tmp_col, q)
  group = rep(NA, length(tmp_col))
  group[tmp_col < cutoffs[1]] = "low"
  group[tmp_col > cutoffs[2]] = "high"
  clinical_data$group = group
  clinical_data = clinical_data |> filter(time < 365 * 5)
  fit = survfit(Surv(time, status) ~ group, data=clinical_data)
  ggsurvplot(fit, data=clinical_data, pval=TRUE, risk.table = FALSE)
}

lapply(c("LUAD", "LUSC"), function(type) {
  lapply(c("sig_neutro", "sig_nan", "sig_tan", "sig_b_cell_follicular", "sig_neutro2", "sig_neutro3"), function(sig) {
    plot_kaplan_meyer(tcga_clin_and_scores |> filter(type == !!type), sig) + ggtitle(sprintf("%s (%s)", sig, type))
  })
})


```

## Responder vs. Non-responder

```{r, fig.width=6, fig.height=6}
glm_res = lapply(colnames(select(ici_clin_and_scores, starts_with("sig_"), starts_with("ct_"))), function(col) {
  print(col)
  res = summary(glm(as.formula(sprintf("response ~ `%s` + dataset", col)), family="binomial", data=ici_clin_and_scores |> mutate(response = factor(response))))
  res$coefficients |> as_tibble(rownames="var") |> filter(grepl( col, var)) |> mutate(var=col)
}) |> bind_rows() |> mutate(sig_code = sig_codes(`Pr(>|z|)`)) |> arrange(-Estimate) |> mutate(var=factor(var, levels=unique(var)))

glm_res |> ggplot(aes(x=var, y=Estimate)) + geom_bar(aes(fill=sig_code), stat="identity") + scale_fill_brewer(palette= "Blues", direction=-1)  + theme_cowplot()+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + geom_hline(yintercept = 0, color="black")
```

```{r, fig.width=18, fig.height=7}
ici_clin_and_scores |> select(type, response, dataset, starts_with("sig_"), starts_with("ct_")) |>
    pivot_longer(-c(type, response, dataset), names_to="signature", values_to="score") |>
    filter(signature %in% (glm_res |> filter(`Pr(>|z|)` < 0.05) |> pull(var))) |> 
    ggplot(aes(x=response, y=score)) + 
    geom_boxplot(outlier.alpha=0, width=.2) + 
    geom_quasirandom(aes(color=type, shape=dataset)) + 
    stat_compare_means(method="wilcox.test") +
    theme_cowplot() + 
    background_grid() + 
    facet_wrap(~signature, ncol = 7, scales="free_y") + 
    ggtitle(sprintf("Scoring function: %s", params$scoring_function)) + 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

```{r, fig.width=30, fig.height=30}
ici_clin_and_scores_long = ici_clin_and_scores |> select(sample, type, response, starts_with("sig_"), starts_with("ct_")) |>
    pivot_longer(-c(sample, type, response), names_to="signature", values_to="score") |> 
    filter(signature != "sig_n_to_t_ratio")

corr_df = ici_clin_and_scores_long |> inner_join(ici_clin_and_scores_long, by=c("sample", "type", "response"))

corr_df |>  
  filter(grepl("sig_", signature.x), grepl("sig_", signature.y)) |> 
  ggplot(aes(x=score.x, y=score.y)) + geom_smooth(method="lm", color="black") + geom_point(aes(color=response)) + stat_cor() + theme_cowplot() + background_grid() + scale_color_brewer(palette="Dark2") + facet_grid(signature.x ~ signature.y)
```

```{r, fig.width=8, fig.height=8}
corr_df2 = corr_df |> group_by(signature.x, signature.y) |> group_modify(function(df, label) { tmp_cor = cor.test(df$score.x, df$score.y); tibble_row(cor = tmp_cor$estimate, pvalue = tmp_cor$p.value) })
corr_mat = corr_df2 |> select(-pvalue) |> pivot_wider(names_from=signature.y, values_from="cor") |> column_to_rownames("signature.x") |> as.matrix()
corrplot(corr_mat)
```

```{r}
ici_clin_and_scores |> 
  ggplot(aes(y=sig_lcam_hi, x=sig_neutro)) + geom_smooth(method="lm", color="black") + geom_point(aes(color=response)) + stat_cor() + theme_cowplot() + background_grid() + scale_color_brewer(palette="Dark2")

ici_clin_and_scores |> 
  ggplot(aes(y=sig_lcam_hi, x=sig_nan)) + geom_smooth(method="lm", color="black") + geom_point(aes(color=response)) + stat_cor() + theme_cowplot() + background_grid() + scale_color_brewer(palette="Dark2")

ici_clin_and_scores |> 
  ggplot(aes(y=sig_lcam_hi, x=sig_tan)) + geom_smooth(method="lm", color="black") + geom_point(aes(color=response)) + stat_cor() + theme_cowplot() + background_grid() + scale_color_brewer(palette="Dark2")

ici_clin_and_scores |> 
  ggplot(aes(y=sig_lcam_hi, x=sig_neutro2)) + geom_smooth(method="lm", color="black") + geom_point(aes(color=response)) + stat_cor() + theme_cowplot() + background_grid() + scale_color_brewer(palette="Dark2")

ici_clin_and_scores |> 
  ggplot(aes(y=sig_lcam_hi, x=sig_neutro3)) + geom_smooth(method="lm", color="black") + geom_point(aes(color=response)) + stat_cor() + theme_cowplot() + background_grid() + scale_color_brewer(palette="Dark2")

```

## Does combining the signatures lead to a better quality of fit? 

(GLM with logistic regression)

```{r}
tmp_df = ici_clin_and_scores
tmp_df$response = as.factor(tmp_df$response)


mod = glm(response ~ sig_neutro + sig_lcam_hi + dataset, data=tmp_df, family=binomial)
summary(mod)

mod2 = glm(response ~ sig_neutro + dataset, data=tmp_df, family=binomial)
summary(mod2)

mod3 = glm(response ~ sig_lcam_hi + dataset, data=tmp_df, family=binomial)
summary(mod3)

# mod4 = glm(response ~ sig_tan + sig_lcam_hi + dataset, data=tmp_df, family=binomial)
# summary(mod4)


```