{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import itertools\n",
    "from tqdm import trange, tqdm\n",
    "import scipy.sparse\n",
    "import numpy.testing as npt\n",
    "from scanpy_helpers.integration import (\n",
    "    normalize_by_gene_length,\n",
    "    sanitize_adata,\n",
    "    validate_adata,\n",
    "    add_doublet_annotation,\n",
    "    undo_log_norm,\n",
    "    remap_gene_symbols,\n",
    "    drop_duplicated_genes,\n",
    "    aggregate_duplicate_gene_symbols,\n",
    "    merge_datasets,\n",
    "    MANDATORY_COLS,\n",
    ")\n",
    "from threadpoolctl import threadpool_limits\n",
    "from tqdm.contrib.concurrent import process_map\n",
    "import mygene\n",
    "from operator import and_\n",
    "from functools import reduce\n",
    "import pandas as pd\n",
    "import anndata\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<threadpoolctl.threadpool_limits at 0x7f9c08424070>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "threadpool_limits(8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(figsize=(5, 5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_table = pd.read_csv(\"../../tables/samplesheet_scrnaseq_preprocessing.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>input_adata</th>\n",
       "      <th>min_counts</th>\n",
       "      <th>max_counts</th>\n",
       "      <th>min_genes</th>\n",
       "      <th>max_genes</th>\n",
       "      <th>max_pct_mito</th>\n",
       "      <th>batch_key</th>\n",
       "      <th>run_solo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Adams_Kaminski_2020_COPD</td>\n",
       "      <td>data/10_public_datasets/Adams_Kaminski_2020_CO...</td>\n",
       "      <td>1000</td>\n",
       "      <td>35000</td>\n",
       "      <td>500</td>\n",
       "      <td>10000</td>\n",
       "      <td>20</td>\n",
       "      <td>sample</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Chen_Zhang_2020_NSCLC</td>\n",
       "      <td>data/10_public_datasets/Chen_Zhang_2020_NSCLC/...</td>\n",
       "      <td>600</td>\n",
       "      <td>30000</td>\n",
       "      <td>250</td>\n",
       "      <td>10000</td>\n",
       "      <td>20</td>\n",
       "      <td>sample</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Goveia_Carmeliet_2020_NSCLC</td>\n",
       "      <td>data/10_public_datasets/Goveia_Carmeliet_2020_...</td>\n",
       "      <td>600</td>\n",
       "      <td>30000</td>\n",
       "      <td>250</td>\n",
       "      <td>10000</td>\n",
       "      <td>20</td>\n",
       "      <td>sample</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Guo_Zhang_2018_NSCLC</td>\n",
       "      <td>data/10_public_datasets/Guo_Zhang_2018_NSCLC/h...</td>\n",
       "      <td>20000</td>\n",
       "      <td>3000000</td>\n",
       "      <td>1000</td>\n",
       "      <td>20000</td>\n",
       "      <td>20</td>\n",
       "      <td>sample</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Habermann_Kropski_2020_pulmonary-fibrosis</td>\n",
       "      <td>data/10_public_datasets/Habermann_Kropski_2020...</td>\n",
       "      <td>600</td>\n",
       "      <td>30000</td>\n",
       "      <td>200</td>\n",
       "      <td>10000</td>\n",
       "      <td>20</td>\n",
       "      <td>sample</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>He_Fan_2021_LUAD</td>\n",
       "      <td>data/10_public_datasets/Kim_Lee_2020_LUAD/h5ad...</td>\n",
       "      <td>1000</td>\n",
       "      <td>35000</td>\n",
       "      <td>300</td>\n",
       "      <td>10000</td>\n",
       "      <td>20</td>\n",
       "      <td>sample</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Kim_Lee_2020_LUAD</td>\n",
       "      <td>data/10_public_datasets/He_Fan_2021_LUAD/h5ad_...</td>\n",
       "      <td>600</td>\n",
       "      <td>30000</td>\n",
       "      <td>250</td>\n",
       "      <td>10000</td>\n",
       "      <td>20</td>\n",
       "      <td>sample</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Lambrechts_2018_LUAD_6149v1</td>\n",
       "      <td>data/10_public_datasets/Lambrechts_2018_LUAD/E...</td>\n",
       "      <td>600</td>\n",
       "      <td>30000</td>\n",
       "      <td>200</td>\n",
       "      <td>10000</td>\n",
       "      <td>15</td>\n",
       "      <td>sample</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Lambrechts_2018_LUAD_6149v2</td>\n",
       "      <td>data/10_public_datasets/Lambrechts_2018_LUAD/E...</td>\n",
       "      <td>600</td>\n",
       "      <td>30000</td>\n",
       "      <td>250</td>\n",
       "      <td>10000</td>\n",
       "      <td>20</td>\n",
       "      <td>sample</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Lambrechts_2018_LUAD_6653</td>\n",
       "      <td>data/10_public_datasets/Lambrechts_2018_LUAD/E...</td>\n",
       "      <td>1200</td>\n",
       "      <td>40000</td>\n",
       "      <td>250</td>\n",
       "      <td>10000</td>\n",
       "      <td>20</td>\n",
       "      <td>sample</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Laughney_Massague_2020_NSCLC</td>\n",
       "      <td>data/10_public_datasets/Laughney_Massague_2020...</td>\n",
       "      <td>1800</td>\n",
       "      <td>40000</td>\n",
       "      <td>500</td>\n",
       "      <td>10000</td>\n",
       "      <td>20</td>\n",
       "      <td>sample</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Madissoon_Meyer_2020_pulmonary-fibrosis</td>\n",
       "      <td>data/10_public_datasets/Madissoon_Meyer_2020_p...</td>\n",
       "      <td>600</td>\n",
       "      <td>30000</td>\n",
       "      <td>300</td>\n",
       "      <td>10000</td>\n",
       "      <td>20</td>\n",
       "      <td>sample</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Maier_Merad_2020_NSCLC</td>\n",
       "      <td>data/10_public_datasets/Maier_Merad_2020_NSCLC...</td>\n",
       "      <td>1000</td>\n",
       "      <td>30000</td>\n",
       "      <td>400</td>\n",
       "      <td>10000</td>\n",
       "      <td>15</td>\n",
       "      <td>sample</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Maynard_Bivona_2020_NSCLC</td>\n",
       "      <td>data/10_public_datasets/Maynard_Bivona_2020_NS...</td>\n",
       "      <td>20000</td>\n",
       "      <td>20000000</td>\n",
       "      <td>600</td>\n",
       "      <td>20000</td>\n",
       "      <td>30</td>\n",
       "      <td>sample</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Mayr_Schiller_2020_pulmonary-fibrosis</td>\n",
       "      <td>data/10_public_datasets/Mayr_Schiller_2020_pul...</td>\n",
       "      <td>600</td>\n",
       "      <td>30000</td>\n",
       "      <td>250</td>\n",
       "      <td>10000</td>\n",
       "      <td>10</td>\n",
       "      <td>sample</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Reyfman_Misharin_2018_pulmonary-fibrosis</td>\n",
       "      <td>data/10_public_datasets/Reyfman_Misharin_2018_...</td>\n",
       "      <td>1000</td>\n",
       "      <td>30000</td>\n",
       "      <td>250</td>\n",
       "      <td>10000</td>\n",
       "      <td>20</td>\n",
       "      <td>sample</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Travaglini_Krasnow_2020_Lung_10x</td>\n",
       "      <td>data/10_public_datasets/Travaglini_Krasnow_202...</td>\n",
       "      <td>1000</td>\n",
       "      <td>30000</td>\n",
       "      <td>500</td>\n",
       "      <td>10000</td>\n",
       "      <td>0</td>\n",
       "      <td>sample</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Travaglini_Krasnow_2020_Lung_SS2</td>\n",
       "      <td>data/10_public_datasets/Travaglini_Krasnow_202...</td>\n",
       "      <td>20000</td>\n",
       "      <td>6000000</td>\n",
       "      <td>600</td>\n",
       "      <td>20000</td>\n",
       "      <td>30</td>\n",
       "      <td>sample</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>UKIM-V</td>\n",
       "      <td>data/11_own_datasets/batch1_3patients/h5ad_raw...</td>\n",
       "      <td>2000</td>\n",
       "      <td>100000</td>\n",
       "      <td>200</td>\n",
       "      <td>8000</td>\n",
       "      <td>30</td>\n",
       "      <td>sample</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Vieira_Teichmann_2019_asthma</td>\n",
       "      <td>data/10_public_datasets/Vieira_Teichmann_2019_...</td>\n",
       "      <td>600</td>\n",
       "      <td>30000</td>\n",
       "      <td>200</td>\n",
       "      <td>10000</td>\n",
       "      <td>20</td>\n",
       "      <td>sample</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Wu_Zhou_2021_NSCLC</td>\n",
       "      <td>data/10_public_datasets/Wu_Zhou_2021_NSCLC/h5a...</td>\n",
       "      <td>600</td>\n",
       "      <td>30000</td>\n",
       "      <td>300</td>\n",
       "      <td>10000</td>\n",
       "      <td>30</td>\n",
       "      <td>sample</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Zilionis_Klein_2019_NSCLC</td>\n",
       "      <td>data/10_public_datasets/Zilionis_Klein_2019_NS...</td>\n",
       "      <td>600</td>\n",
       "      <td>30000</td>\n",
       "      <td>200</td>\n",
       "      <td>10000</td>\n",
       "      <td>20</td>\n",
       "      <td>sample</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           id  \\\n",
       "0                    Adams_Kaminski_2020_COPD   \n",
       "1                       Chen_Zhang_2020_NSCLC   \n",
       "2                 Goveia_Carmeliet_2020_NSCLC   \n",
       "3                        Guo_Zhang_2018_NSCLC   \n",
       "4   Habermann_Kropski_2020_pulmonary-fibrosis   \n",
       "5                            He_Fan_2021_LUAD   \n",
       "6                           Kim_Lee_2020_LUAD   \n",
       "7                 Lambrechts_2018_LUAD_6149v1   \n",
       "8                 Lambrechts_2018_LUAD_6149v2   \n",
       "9                   Lambrechts_2018_LUAD_6653   \n",
       "10               Laughney_Massague_2020_NSCLC   \n",
       "11    Madissoon_Meyer_2020_pulmonary-fibrosis   \n",
       "12                     Maier_Merad_2020_NSCLC   \n",
       "13                  Maynard_Bivona_2020_NSCLC   \n",
       "14      Mayr_Schiller_2020_pulmonary-fibrosis   \n",
       "15   Reyfman_Misharin_2018_pulmonary-fibrosis   \n",
       "16           Travaglini_Krasnow_2020_Lung_10x   \n",
       "17           Travaglini_Krasnow_2020_Lung_SS2   \n",
       "18                                     UKIM-V   \n",
       "19               Vieira_Teichmann_2019_asthma   \n",
       "20                         Wu_Zhou_2021_NSCLC   \n",
       "21                  Zilionis_Klein_2019_NSCLC   \n",
       "\n",
       "                                          input_adata  min_counts  max_counts  \\\n",
       "0   data/10_public_datasets/Adams_Kaminski_2020_CO...        1000       35000   \n",
       "1   data/10_public_datasets/Chen_Zhang_2020_NSCLC/...         600       30000   \n",
       "2   data/10_public_datasets/Goveia_Carmeliet_2020_...         600       30000   \n",
       "3   data/10_public_datasets/Guo_Zhang_2018_NSCLC/h...       20000     3000000   \n",
       "4   data/10_public_datasets/Habermann_Kropski_2020...         600       30000   \n",
       "5   data/10_public_datasets/Kim_Lee_2020_LUAD/h5ad...        1000       35000   \n",
       "6   data/10_public_datasets/He_Fan_2021_LUAD/h5ad_...         600       30000   \n",
       "7   data/10_public_datasets/Lambrechts_2018_LUAD/E...         600       30000   \n",
       "8   data/10_public_datasets/Lambrechts_2018_LUAD/E...         600       30000   \n",
       "9   data/10_public_datasets/Lambrechts_2018_LUAD/E...        1200       40000   \n",
       "10  data/10_public_datasets/Laughney_Massague_2020...        1800       40000   \n",
       "11  data/10_public_datasets/Madissoon_Meyer_2020_p...         600       30000   \n",
       "12  data/10_public_datasets/Maier_Merad_2020_NSCLC...        1000       30000   \n",
       "13  data/10_public_datasets/Maynard_Bivona_2020_NS...       20000    20000000   \n",
       "14  data/10_public_datasets/Mayr_Schiller_2020_pul...         600       30000   \n",
       "15  data/10_public_datasets/Reyfman_Misharin_2018_...        1000       30000   \n",
       "16  data/10_public_datasets/Travaglini_Krasnow_202...        1000       30000   \n",
       "17  data/10_public_datasets/Travaglini_Krasnow_202...       20000     6000000   \n",
       "18  data/11_own_datasets/batch1_3patients/h5ad_raw...        2000      100000   \n",
       "19  data/10_public_datasets/Vieira_Teichmann_2019_...         600       30000   \n",
       "20  data/10_public_datasets/Wu_Zhou_2021_NSCLC/h5a...         600       30000   \n",
       "21  data/10_public_datasets/Zilionis_Klein_2019_NS...         600       30000   \n",
       "\n",
       "    min_genes  max_genes  max_pct_mito batch_key  run_solo  \n",
       "0         500      10000            20    sample      True  \n",
       "1         250      10000            20    sample      True  \n",
       "2         250      10000            20    sample      True  \n",
       "3        1000      20000            20    sample     False  \n",
       "4         200      10000            20    sample      True  \n",
       "5         300      10000            20    sample      True  \n",
       "6         250      10000            20    sample      True  \n",
       "7         200      10000            15    sample      True  \n",
       "8         250      10000            20    sample      True  \n",
       "9         250      10000            20    sample      True  \n",
       "10        500      10000            20    sample      True  \n",
       "11        300      10000            20    sample      True  \n",
       "12        400      10000            15    sample      True  \n",
       "13        600      20000            30    sample     False  \n",
       "14        250      10000            10    sample      True  \n",
       "15        250      10000            20    sample      True  \n",
       "16        500      10000             0    sample      True  \n",
       "17        600      20000            30    sample     False  \n",
       "18        200       8000            30    sample      True  \n",
       "19        200      10000            20    sample      True  \n",
       "20        300      10000            30    sample      True  \n",
       "21        200      10000            20    sample      True  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 22/22 [00:34<00:00,  1.55s/it]\n"
     ]
    }
   ],
   "source": [
    "datasets = {\n",
    "    dataset_id: sc.read_h5ad(\n",
    "        f\"../../data/20_qc_norm_scrnaseq/01_qc_and_filtering/{dataset_id}/{dataset_id}.qc.h5ad\"\n",
    "    )\n",
    "    for dataset_id in tqdm(dataset_table[\"id\"])\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dataset-specific filtering and metadata fixes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/sturm/.conda/envs/pircher-sc-integrate2/lib/python3.8/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "datasets[\"Adams_Kaminski_2020_COPD\"].obs[\"origin\"] = \"normal\"\n",
    "datasets[\"Adams_Kaminski_2020_COPD\"].obs[\"sex\"] = \"nan\"\n",
    "datasets[\"Adams_Kaminski_2020_COPD\"] = datasets[\"Adams_Kaminski_2020_COPD\"][\n",
    "    datasets[\"Adams_Kaminski_2020_COPD\"].obs[\"condition\"] != \"IPF\", :\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# No modifications necessary for Chen_Zhang"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets[\"Goveia_Carmeliet_2020_NSCLC\"] = datasets[\"Goveia_Carmeliet_2020_NSCLC\"][\n",
    "    datasets[\"Goveia_Carmeliet_2020_NSCLC\"].obs[\"condition\"] != \"LLCC\"\n",
    "].copy()\n",
    "datasets[\"Goveia_Carmeliet_2020_NSCLC\"].obs[\"sex\"] = \"nan\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets[\"Guo_Zhang_2018_NSCLC\"] = datasets[\"Guo_Zhang_2018_NSCLC\"][\n",
    "    datasets[\"Guo_Zhang_2018_NSCLC\"].obs[\"tissue\"] != \"blood\"\n",
    "].copy()\n",
    "datasets[\"Guo_Zhang_2018_NSCLC\"] = normalize_by_gene_length(\n",
    "    datasets[\"Guo_Zhang_2018_NSCLC\"]\n",
    ")\n",
    "datasets[\"Guo_Zhang_2018_NSCLC\"].obs[\"sex\"] = \"nan\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets[\"Habermann_Kropski_2020_pulmonary-fibrosis\"].obs[\"sex\"] = [\n",
    "    {\"M\": \"male\", \"F\": \"female\", \"Unknown\": \"nan\"}[s]\n",
    "    for s in datasets[\"Habermann_Kropski_2020_pulmonary-fibrosis\"].obs[\"sex\"]\n",
    "]\n",
    "datasets[\"Habermann_Kropski_2020_pulmonary-fibrosis\"] = datasets[\n",
    "    \"Habermann_Kropski_2020_pulmonary-fibrosis\"\n",
    "][\n",
    "    datasets[\"Habermann_Kropski_2020_pulmonary-fibrosis\"].obs[\"condition\"]\n",
    "    == \"healthy_control\",\n",
    "    :,\n",
    "].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# No modifications necessary for He_Fan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets[\"Maynard_Bivona_2020_NSCLC\"] = normalize_by_gene_length(\n",
    "    datasets[\"Maynard_Bivona_2020_NSCLC\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets[\"Laughney_Massague_2020_NSCLC\"].obs[\"sex\"] = \"nan\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets[\"Maier_Merad_2020_NSCLC\"].obs[\"sex\"] = \"nan\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets[\"Madissoon_Meyer_2020_pulmonary-fibrosis\"].obs[\"tissue\"] = \"lung\"\n",
    "datasets[\"Madissoon_Meyer_2020_pulmonary-fibrosis\"].obs[\"origin\"] = \"normal\"\n",
    "datasets[\"Madissoon_Meyer_2020_pulmonary-fibrosis\"].obs[\"condition\"] = \"healthy_control\"\n",
    "datasets[\"Madissoon_Meyer_2020_pulmonary-fibrosis\"].obs[\"sex\"] = \"nan\"\n",
    "datasets[\"Madissoon_Meyer_2020_pulmonary-fibrosis\"].X.data = np.rint(\n",
    "    datasets[\"Madissoon_Meyer_2020_pulmonary-fibrosis\"].X.data\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets[\"Mayr_Schiller_2020_pulmonary-fibrosis\"].obs[\"sex\"] = \"nan\"\n",
    "datasets[\"Mayr_Schiller_2020_pulmonary-fibrosis\"] = datasets[\n",
    "    \"Mayr_Schiller_2020_pulmonary-fibrosis\"\n",
    "][\n",
    "    datasets[\"Mayr_Schiller_2020_pulmonary-fibrosis\"].obs[\"condition\"]\n",
    "    == \"healthy_control\",\n",
    "    :,\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets[\"Reyfman_Misharin_2018_pulmonary-fibrosis\"].obs[\"sex\"] = \"nan\"\n",
    "datasets[\"Reyfman_Misharin_2018_pulmonary-fibrosis\"] = datasets[\n",
    "    \"Reyfman_Misharin_2018_pulmonary-fibrosis\"\n",
    "][\n",
    "    datasets[\"Reyfman_Misharin_2018_pulmonary-fibrosis\"].obs[\"condition\"]\n",
    "    == \"healthy_control\",\n",
    "    :,\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets[\"Travaglini_Krasnow_2020_Lung_SS2\"] = datasets[\n",
    "    \"Travaglini_Krasnow_2020_Lung_SS2\"\n",
    "][datasets[\"Travaglini_Krasnow_2020_Lung_SS2\"].obs[\"tissue\"] == \"lung\", :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.obs` of view, copying.\n"
     ]
    }
   ],
   "source": [
    "datasets[\"Zilionis_Klein_2019_NSCLC\"] = datasets[\"Zilionis_Klein_2019_NSCLC\"][\n",
    "    datasets[\"Zilionis_Klein_2019_NSCLC\"].obs[\"tissue\"] == \"lung\", :\n",
    "]\n",
    "datasets[\"Zilionis_Klein_2019_NSCLC\"].obs[\"sex\"] = [\n",
    "    {\"M\": \"male\", \"F\": \"female\", \"Unknown\": \"nan\"}[s]\n",
    "    for s in datasets[\"Zilionis_Klein_2019_NSCLC\"].obs[\"sex\"]\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### make patients unique across datasets\n",
    "\n",
    "Except for the two Travaglini variants - they are the same patients profiled with different platforms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.obs` of view, copying.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Trying to set attribute `.obs` of view, copying.\n"
     ]
    }
   ],
   "source": [
    "for dataset_id, adata in datasets.items():\n",
    "    adata.obs[\"dataset\"] = dataset_id\n",
    "    adata.obs[\"patient\"] = [\n",
    "        f'{dataset.replace(\"_10x\", \"\").replace(\"_SS2\", \"\")}_{patient}'\n",
    "        for dataset, patient in zip(adata.obs[\"dataset\"], adata.obs[\"patient\"])\n",
    "    ]\n",
    "    datasets[dataset_id] = adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Validate data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating Adams_Kaminski_2020_COPD\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'patient' as categorical\n",
      "... storing 'sample' as categorical\n",
      "... storing 'origin' as categorical\n",
      "... storing 'sex' as categorical\n",
      "... storing 'dataset' as categorical\n",
      "... storing 'dataset' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating Chen_Zhang_2020_NSCLC\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'sample' as categorical\n",
      "... storing 'patient' as categorical\n",
      "... storing 'patient' as categorical\n",
      "... storing 'sample' as categorical\n",
      "... storing 'sex' as categorical\n",
      "... storing 'dataset' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating Goveia_Carmeliet_2020_NSCLC\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'patient' as categorical\n",
      "... storing 'sample' as categorical\n",
      "... storing 'sex' as categorical\n",
      "... storing 'dataset' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating Guo_Zhang_2018_NSCLC\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'sample' as categorical\n",
      "... storing 'sex' as categorical\n",
      "... storing 'patient' as categorical\n",
      "... storing 'dataset' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating Habermann_Kropski_2020_pulmonary-fibrosis\n",
      "Validating He_Fan_2021_LUAD\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'sample' as categorical\n",
      "... storing 'patient' as categorical\n",
      "... storing 'dataset' as categorical\n",
      "... storing 'dataset' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating Kim_Lee_2020_LUAD\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'sample' as categorical\n",
      "... storing 'patient' as categorical\n",
      "... storing 'dataset' as categorical\n",
      "... storing 'sample' as categorical\n",
      "... storing 'patient' as categorical\n",
      "... storing 'dataset' as categorical\n",
      "... storing 'sample' as categorical\n",
      "... storing 'patient' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating Lambrechts_2018_LUAD_6149v1\n",
      "Validating Lambrechts_2018_LUAD_6149v2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'dataset' as categorical\n",
      "... storing 'sample' as categorical\n",
      "... storing 'patient' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating Lambrechts_2018_LUAD_6653\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'sample' as categorical\n",
      "... storing 'patient' as categorical\n",
      "... storing 'sex' as categorical\n",
      "... storing 'dataset' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating Laughney_Massague_2020_NSCLC\n",
      "Validating Madissoon_Meyer_2020_pulmonary-fibrosis\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'patient' as categorical\n",
      "... storing 'sample' as categorical\n",
      "... storing 'tissue' as categorical\n",
      "... storing 'origin' as categorical\n",
      "... storing 'condition' as categorical\n",
      "... storing 'sex' as categorical\n",
      "... storing 'dataset' as categorical\n",
      "... storing 'sample' as categorical\n",
      "... storing 'patient' as categorical\n",
      "... storing 'sex' as categorical\n",
      "... storing 'dataset' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating Maier_Merad_2020_NSCLC\n",
      "Validating Maynard_Bivona_2020_NSCLC\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'sample' as categorical\n",
      "... storing 'patient' as categorical\n",
      "... storing 'dataset' as categorical\n",
      "... storing 'sample' as categorical\n",
      "... storing 'patient' as categorical\n",
      "... storing 'dataset' as categorical\n",
      "... storing 'sex' as categorical\n",
      "... storing 'dataset' as categorical\n",
      "... storing 'sample' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating Mayr_Schiller_2020_pulmonary-fibrosis\n",
      "Validating Reyfman_Misharin_2018_pulmonary-fibrosis\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'patient' as categorical\n",
      "... storing 'sex' as categorical\n",
      "... storing 'sample' as categorical\n",
      "... storing 'patient' as categorical\n",
      "... storing 'dataset' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating Travaglini_Krasnow_2020_Lung_10x\n",
      "Validating Travaglini_Krasnow_2020_Lung_SS2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'patient' as categorical\n",
      "... storing 'sample' as categorical\n",
      "... storing 'dataset' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating UKIM-V\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'sample' as categorical\n",
      "... storing 'dataset' as categorical\n",
      "... storing 'patient' as categorical\n",
      "... storing 'sample' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating Vieira_Teichmann_2019_asthma\n",
      "Validating Wu_Zhou_2021_NSCLC\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'patient' as categorical\n",
      "... storing 'dataset' as categorical\n",
      "... storing 'sample' as categorical\n",
      "... storing 'sex' as categorical\n",
      "... storing 'patient' as categorical\n",
      "... storing 'dataset' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Validating Zilionis_Klein_2019_NSCLC\n"
     ]
    }
   ],
   "source": [
    "for dataset_id, adata in datasets.items():\n",
    "    print(f\"Validating {dataset_id}\")\n",
    "    sanitize_adata(adata)\n",
    "    validate_adata(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene identifier remapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "07c6da5d31d04b09a8ce9507418d5d14",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=22.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "querying 1-1000...done.\n",
      "querying 1-1000...querying 1001-2000...querying 1-1000...querying 1-1000...done.\n",
      "querying 1001-2000...done.\n",
      "done.\n",
      "querying 1-1000...done.\n",
      "querying 1001-2000...querying 2001-3000...querying 1001-2000...done.\n",
      "done.\n",
      "querying 2001-3000...done.\n",
      "querying 1001-2000...done.\n",
      "done.\n",
      "querying 2001-3000...done.\n",
      "querying 2001-3000...querying 3001-4000...done.\n",
      "querying 3001-4000...querying 2001-3000...done.\n",
      "done.\n",
      "done.\n",
      "querying 3001-4000...querying 4001-5000...querying 3001-4000...done.\n",
      "done.\n",
      "querying 4001-5000...querying 3001-4000...querying 1-1000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 4001-5000...done.\n",
      "done.\n",
      "querying 5001-6000...querying 4001-5000...querying 5001-6000...querying 1001-2000...querying 4001-5000...querying 1-1000...\n",
      "done.querying 1-1000...querying 5001-6000...querying 1-1000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 1-1000...querying 6001-7000...querying 6001-7000...querying 5001-6000...querying 5001-6000...querying 2001-3000...done.\n",
      "done.\n",
      "querying 1-1000...done.\n",
      "done.\n",
      "querying 1001-2000...querying 1001-2000...querying 1001-2000...done.\n",
      "querying 6001-7000...done.\n",
      "querying 6001-7000...done.\n",
      "querying 1-1000...\n",
      "querying 7001-8000...querying 1-1000...querying 6001-7000...done.done.\n",
      "done.\n",
      "done.\n",
      "querying 1001-2000...querying 7001-8000...querying 3001-4000...querying 7001-8000...\n",
      "querying 1-1000...done.\n",
      "done.\n",
      "querying 1-1000...done.querying 8001-9000...done.\n",
      "done.\n",
      "querying 7001-8000...done.\n",
      "querying 2001-3000...querying 1001-2000...querying 1001-2000...querying 4001-5000...querying 1-1000...done.\n",
      "done.\n",
      "done.\n",
      "querying 1-1000...done.\n",
      "done.\n",
      "done.\n",
      "querying 2001-3000...done.\n",
      "querying 2001-3000...querying 8001-9000...querying 8001-9000...querying 2001-3000...querying 7001-8000...querying 9001-10000...done.\n",
      "querying 1-1000...querying 1001-2000...querying 1-1000...querying 1-1000...done.\n",
      "done.\n",
      "querying 8001-9000...querying 1001-2000...done.\n",
      "done.\n",
      "done.\n",
      "querying 1001-2000...querying 1001-2000...querying 1001-2000...done.done.\n",
      "\n",
      "done.done.\n",
      "\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 1-1000...querying 1-1000...done.done.\n",
      "\n",
      "querying 3001-4000...querying 2001-3000...done.done.\n",
      "\n",
      "done.\n",
      "querying 1001-2000...querying 5001-6000...querying 3001-4000...querying 2001-3000...querying 2001-3000...querying 9001-10000...querying 1001-2000...querying 2001-3000...querying 3001-4000...querying 9001-10000...querying 8001-9000...querying 3001-4000...done.done.done.\n",
      "\n",
      "\n",
      "done.\n",
      "done.done.\n",
      "\n",
      "done.\n",
      "done.\n",
      "querying 2001-3000...querying 1001-2000...querying 1001-2000...querying 10001-11000...querying 2001-3000...querying 2001-3000...querying 9001-10000...querying 1001-2000...done.done.\n",
      "\n",
      "done.\n",
      "querying 4001-5000...querying 4001-5000...querying 10001-11000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.done.\n",
      "\n",
      "done.\n",
      "querying 3001-4000...querying 3001-4000...querying 2001-3000...querying 3001-4000...querying 4001-5000...querying 9001-10000...done.\n",
      "querying 4001-5000...done.\n",
      "done.\n",
      "querying 6001-7000...querying 2001-3000...querying 3001-4000...done.\n",
      "done.\n",
      "done.done.\n",
      "\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 2001-3000...done.\n",
      "querying 2001-3000...querying 3001-4000...querying 11001-12000...querying 3001-4000...done.\n",
      "querying 3001-4000...done.\n",
      "querying 5001-6000...querying 4001-5000...done.done.\n",
      "\n",
      "querying 4001-5000...querying 3001-4000...querying 10001-11000...querying 11001-12000...done.\n",
      "done.done.\n",
      "\n",
      "querying 4001-5000...querying 10001-11000...querying 4001-5000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 5001-6000...querying 5001-6000...querying 5001-6000...querying 2001-3000...querying 4001-5000...querying 10001-11000...done.\n",
      "done.\n",
      "done.\n",
      "querying 5001-6000...done.\n",
      "querying 3001-4000...querying 3001-4000...querying 4001-5000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 4001-5000...done.\n",
      "querying 11001-12000...querying 7001-8000...querying 5001-6000...querying 5001-6000...querying 12001-13000...done.\n",
      "querying 6001-7000...done.\n",
      "done.\n",
      "querying 4001-5000...done.done.\n",
      "\n",
      "querying 4001-5000...querying 6001-7000...querying 6001-7000...querying 12001-13000...done.\n",
      "done.done.\n",
      "\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 3001-4000...querying 6001-7000...querying 11001-12000...querying 8001-9000...querying 5001-6000...querying 11001-12000...done.\n",
      "done.\n",
      "done.\n",
      "querying 12001-13000...done.\n",
      "done.\n",
      "querying 13001-14000...done.\n",
      "querying 5001-6000...done.done.\n",
      "\n",
      "querying 5001-6000...done.\n",
      "done.\n",
      "querying 13001-14000...querying 6001-7000...querying 5001-6000...querying 3001-4000...querying 4001-5000...querying 4001-5000...querying 7001-8000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 6001-7000...querying 6001-7000...querying 13001-14000...querying 7001-8000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 12001-13000...done.\n",
      "done.\n",
      "querying 4001-5000...done.\n",
      "querying 5001-6000...querying 12001-13000...querying 7001-8000...querying 8001-9000...querying 7001-8000...querying 6001-7000...querying 5001-6000...querying 6001-7000...querying 7001-8000...done.\n",
      "done.\n",
      "done.\n",
      "querying 14001-15000...done.done.\n",
      "\n",
      "querying 5001-6000...done.\n",
      "done.\n",
      "querying 14001-15000...querying 14001-15000...querying 7001-8000...done.\n",
      "querying 8001-9000...querying 13001-14000...done.\n",
      "done.\n",
      "done.\n",
      "querying 6001-7000...querying 15001-16000...querying 13001-14000...querying 6001-7000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 9001-10000...done.\n",
      "querying 7001-8000...querying 5001-6000...done.\n",
      "querying 6001-7000...done.\n",
      "querying 8001-9000...done.\n",
      "done.\n",
      "querying 7001-8000...done.\n",
      "done.\n",
      "querying 7001-8000...done.\n",
      "querying 9001-10000...querying 16001-17000...done.\n",
      "done.\n",
      "querying 6001-7000...querying 8001-9000...querying 8001-9000...done.\n",
      "done.\n",
      "querying 9001-10000...querying 14001-15000...done.\n",
      "done.\n",
      "querying 5001-6000...querying 6001-7000...querying 8001-9000...querying 8001-9000...done.\n",
      "done.\n",
      "done.\n",
      "querying 7001-8000...querying 6001-7000...querying 15001-16000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 9001-10000...querying 7001-8000...done.\n",
      "done.\n",
      "done.\n",
      "querying 7001-8000...querying 14001-15000...querying 9001-10000...querying 8001-9000...querying 9001-10000...done.\n",
      "done.\n",
      "done.\n",
      "querying 9001-10000...done.\n",
      "querying 15001-16000...querying 10001-11000...done.\n",
      "querying 8001-9000...querying 10001-11000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 16001-17000...querying 15001-16000...querying 16001-17000...querying 7001-8000...querying 9001-10000...querying 10001-11000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.done.\n",
      "done.\n",
      "\n",
      "done.\n",
      "done.\n",
      "querying 6001-7000...querying 9001-10000...querying 8001-9000...querying 7001-8000...querying 11001-12000...querying 15001-16000...querying 7001-8000...querying 10001-11000...done.\n",
      "done.\n",
      "querying 11001-12000...done.\n",
      "done.\n",
      "querying 17001-18000...querying 10001-11000...querying 17001-18000...querying 8001-9000...done.done.done.\n",
      "\n",
      "\n",
      "done.\n",
      "querying 10001-11000...querying 10001-11000...querying 11001-12000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 16001-17000...done.\n",
      "querying 8001-9000...querying 9001-10000...done.\n",
      "done.\n",
      "querying 9001-10000...querying 12001-13000...done.\n",
      "querying 18001-19000...querying 10001-11000...querying 8001-9000...querying 12001-13000...done.\n",
      "done.\n",
      "done.\n",
      "querying 11001-12000...done.\n",
      "querying 17001-18000...done.\n",
      "querying 18001-19000...done.\n",
      "done.\n",
      "querying 8001-9000...querying 8001-9000...querying 16001-17000...querying 10001-11000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 19001-20000...querying 12001-13000...done.\n",
      "done.\n",
      "done.\n",
      "querying 9001-10000...querying 17001-18000...querying 11001-12000...querying 12001-13000...querying 10001-11000...done.\n",
      "querying 10001-11000...querying 19001-20000...querying 13001-14000...querying 7001-8000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 18001-18450...done.\n",
      "querying 10001-11000...querying 9001-10000...querying 13001-14000...querying 11001-12000...done.\n",
      "done.done.done.\n",
      "\n",
      "\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 11001-12000...querying 20001-21000...querying 14001-15000...querying 9001-10000...querying 18001-19000...querying 13001-14000...querying 8001-9000...done.\n",
      "done.done.\n",
      "\n",
      "done.\n",
      "Finished.\n",
      "4140 input query terms found dup hits:\n",
      "\t[('A1BG', 2), ('A2M', 5), ('A2ML1', 5), ('AAK1', 2), ('AARSD1', 4), ('ABCA9', 2), ('ABCB10', 10), ('\n",
      "2129 input query terms found no hit:\n",
      "done.\n",
      "\t['ABC12-49244600F4.4', 'AC000403.4', 'AC002076.10', 'AC002116.7', 'AC002117.1', 'AC002310.12', 'AC00\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "Re-querying with the following entries: \n",
      "[]\n",
      "Finished.\n",
      "Remapping complete! 17641/18450 genes remapped.\n",
      "querying 11001-12000...querying 9001-10000...querying 11001-12000...done.done.\n",
      "\n",
      "querying 9001-10000...done.done.done.\n",
      "\n",
      "\n",
      "querying 14001-15000...querying 20001-21000...done.\n",
      "querying 11001-12000...querying 10001-11000...querying 13001-14000...done.\n",
      "done.\n",
      "done.\n",
      "querying 11001-12000...done.\n",
      "done.\n",
      "done.\n",
      "querying 12001-13000...done.\n",
      "done.\n",
      "querying 9001-10000...querying 15001-16000...querying 19001-20000...querying 21001-22000...querying 14001-15000...querying 10001-11000...done.\n",
      "querying 12001-13000...querying 15001-16000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 17001-18000...querying 12001-13000...querying 13001-14000...done.\n",
      "querying 14001-15000...querying 10001-11000...querying 12001-13000...done.\n",
      "done.\n",
      "querying 21001-22000...querying 12001-13000...querying 10001-11000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 11001-12000...querying 20001-21000...querying 11001-12000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 12001-13000...done.\n",
      "querying 16001-17000...querying 15001-16000...done.\n",
      "querying 16001-17000...querying 11001-12000...querying 22001-23000...done.\n",
      "querying 18001-19000...done.\n",
      "querying 13001-14000...querying 13001-14000...done.done.\n",
      "\n",
      "done.\n",
      "querying 13001-14000...querying 13001-14000...done.\n",
      "querying 12001-13000...querying 11001-12000...querying 14001-15000...done.\n",
      "done.\n",
      "querying 22001-23000...done.\n",
      "querying 12001-13000...done.\n",
      "done.\n",
      "done.\n",
      "querying 10001-11000...done.\n",
      "querying 15001-16000...done.\n",
      "querying 23001-24000...querying 21001-22000...querying 16001-17000...done.\n",
      "querying 14001-15000...querying 17001-18000...done.\n",
      "done.\n",
      "querying 14001-15000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 12001-13000...done.\n",
      "querying 23001-24000...querying 17001-18000...querying 11001-12000...done.\n",
      "querying 13001-14000...querying 14001-15000...querying 13001-14000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 19001-20000...done.\n",
      "querying 13001-14000...done.\n",
      "querying 15001-16000...querying 24001-25000...querying 18001-19000...done.\n",
      "done.\n",
      "querying 15001-16000...querying 15001-16000...done.\n",
      "done.\n",
      "querying 14001-15000...querying 12001-13000...querying 22001-23000...done.\n",
      "done.\n",
      "querying 11001-12000...done.\n",
      "done.\n",
      "done.\n",
      "querying 12001-13000...querying 16001-17000...querying 17001-18000...querying 16001-17000...querying 19001-20000...done.done.\n",
      "\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 23001-24000...querying 14001-15000...querying 14001-15000...querying 18001-19000...querying 16001-17000...done.done.\n",
      "\n",
      "querying 13001-14000...done.\n",
      "done.\n",
      "querying 12001-13000...querying 16001-17000...done.\n",
      "querying 15001-16000...querying 18001-19000...querying 12001-13000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 20001-21000...done.\n",
      "done.\n",
      "querying 17001-18000...done.\n",
      "done.\n",
      "querying 13001-14000...querying 13001-14000...done.\n",
      "done.\n",
      "querying 14001-15000...querying 17001-18000...querying 14001-15000...querying 24001-25000...querying 19001-20000...querying 20001-21000...done.\n",
      "querying 13001-14000...done.done.\n",
      "\n",
      "done.done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "\n",
      "querying 25001-26000...querying 15001-16000...done.\n",
      "done.\n",
      "querying 13001-14000...querying 15001-16000...querying 17001-18000...querying 14001-15000...querying 15001-16000...querying 17001-18000...done.\n",
      "querying 21001-22000...querying 25001-26000...querying 15001-16000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 18001-19000...querying 24001-25000...querying 18001-19000...done.\n",
      "done.\n",
      "done.\n",
      "querying 14001-15000...querying 16001-17000...done.\n",
      "querying 21001-22000...querying 18001-19000...done.\n",
      "done.\n",
      "done.\n",
      "querying 14001-15000...querying 22001-23000...done.\n",
      "done.\n",
      "querying 14001-15000...querying 26001-27000...done.\n",
      "done.\n",
      "done.done.\n",
      "\n",
      "querying 26001-27000...querying 19001-20000...querying 15001-16000...querying 16001-17000...done.\n",
      "querying 22001-23000...querying 25001-26000...done.\n",
      "done.\n",
      "done.\n",
      "querying 15001-16000...done.\n",
      "querying 16001-17000...done.\n",
      "done.\n",
      "querying 16001-17000...querying 20001-21000...done.\n",
      "querying 16001-17000...done.\n",
      "querying 23001-24000...querying 27001-27717...querying 15001-16000...done.done.\n",
      "\n",
      "querying 27001-28000...done.done.\n",
      "querying 15001-16000...\n",
      "querying 18001-19000...done.\n",
      "done.\n",
      "querying 16001-17000...querying 15001-16000...done.\n",
      "querying 19001-20000...Finished.\n",
      "5248 input query terms found dup hits:\n",
      "\t[('TTLL10', 2), ('DVL1', 2), ('MRPL20', 4), ('SSU72', 10), ('SLC35E2', 2), ('PRKCZ', 2), ('SKI', 5),\n",
      "7463 input query terms found no hit:\n",
      "\t['RP11-34P13.3', 'RP11-34P13.7', 'RP11-34P13.8', 'RP11-34P13.14', 'FO538757.3', 'FO538757.2', 'AP006\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "querying 24001-25000...The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "Re-querying with the following entries: \n",
      "[]\n",
      "Finished.done.\n",
      "\n",
      "Remapping complete! 26661/27717 genes remapped.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 19001-20000...querying 20001-21000...querying 26001-27000...querying 28001-29000...querying 17001-18000...querying 23001-24000...done.\n",
      "done.\n",
      "done.\n",
      "querying 17001-18000...querying 21001-22000...querying 16001-17000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 17001-18000...querying 16001-17000...done.\n",
      "done.\n",
      "done.\n",
      "querying 19001-20000...querying 20001-21000...done.\n",
      "querying 20001-21000...querying 17001-18000...done.\n",
      "done.\n",
      "querying 17001-18000...done.\n",
      "querying 18001-19000...done.\n",
      "querying 16001-16322...querying 29001-30000...done.\n",
      "querying 25001-26000...querying 21001-22000...done.\n",
      "done.\n",
      "done.\n",
      "querying 22001-23000...done.\n",
      "done.\n",
      "querying 24001-25000...done.\n",
      "querying 18001-19000...querying 18001-19000...done.\n",
      "querying 17001-18000...done.\n",
      "done.\n",
      "querying 20001-21000...done.\n",
      "querying 30001-31000...querying 17001-18000...done.\n",
      "querying 17001-18000...querying 27001-28000...querying 18001-19000...querying 19001-20000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 26001-26359...querying 23001-24000...done.\n",
      "querying 18001-19000...Finished.\n",
      "3633 input query terms found dup hits:\n",
      "\t[('A2M', 5), ('AAK1', 2), ('AARSD1', 4), ('ABCA9', 2), ('ABCB10', 10), ('ABCC2', 2), ('ABCC3', 2), (\n",
      "1818 input query terms found no hit:\n",
      "\t['AC002044.4', 'AC002064.4', 'AC002117.1', 'AC002310.12', 'AC002310.7', 'AC002398.11', 'AC002398.12'\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "Re-querying with the following entries: \n",
      "['MIR1254-1', 'hsa-mir-4259']\n",
      "querying 1-2...querying 28001-28863...done.\n",
      "done.done.done.\n",
      "done.done.\n",
      "\n",
      "done.\n",
      "\n",
      "\n",
      "done.done.\n",
      "\n",
      "Finished.\n",
      "5401 input query terms found dup hits:\n",
      "\t[('TTLL10', 2), ('LINC01786', 2), ('DVL1', 2), ('MRPL20', 4), ('LINC01770', 2), ('SSU72', 10), ('PRK\n",
      "6381 input query terms found no hit:\n",
      "\t['AL627309.1', 'AL627309.5', 'AP006222.2', 'AL645608.2', 'AL645608.4', 'AL645608.7', 'AL390719.3', '\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "querying 19001-19681...Re-querying with the following entries: querying 25001-26000...querying 19001-20000...\n",
      "querying 21001-22000...querying 31001-32000...['CYB561D2-1', 'GGT1-1', 'HSPA14-1', 'LINC01238-1', 'TBCE-1']querying 21001-21595...\n",
      "querying 16001-17000...querying 1-5...querying 18001-19000...done.\n",
      "done.done.\n",
      "\n",
      "done.done.done.done.\n",
      "\n",
      "\n",
      "\n",
      "done.\n",
      "done.\n",
      "Finished.querying 22001-23000...\n",
      "querying 19001-20000...1 input query terms found dup hits:\n",
      "\t[('hsa-mir', 2)]\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "Remapping complete! 15327/16322 genes remapped.\n",
      "querying 18001-19000...querying 18001-19000...querying 20001-21000...querying 19001-19201...querying 24001-25000...done.\n",
      "querying 21001-22000...done.\n",
      "done.\n",
      "Finished.\n",
      "5720 input query terms found dup hits:\n",
      "\t[('TTLL10', 2), ('LINC01786', 2), ('DVL1', 2), ('MRPL20', 4), ('LINC01770', 2), ('SSU72', 10), ('PRK\n",
      "7916 input query terms found no hit:\n",
      "\t['AL627309.1', 'AL627309.5', 'AP006222.2', 'AC114498.1', 'AL645608.6', 'AL645608.2', 'AL645608.4', '\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "Re-querying with the following entries: \n",
      "['ARMCX5-GPRASP2-1', 'CYB561D2-1', 'GGT1-1', 'GOLGA8M-1', 'HSPA14-1', 'LINC01238-1', 'TBCE-1']\n",
      "querying 1-7...done.\n",
      "done.\n",
      "done.\n",
      "querying 19001-20000...Finished.\n",
      "4674 input query terms found dup hits:\n",
      "\t[('A1BG', 2), ('A2M', 5), ('A2ML1', 5), ('AADACL2', 4), ('AAK1', 2), ('AARSD1', 4), ('ABCA9', 2), ('\n",
      "3148 input query terms found no hit:\n",
      "\t['AB015752.3', 'ABC12-49244600F4.4', 'ABC13-47488600E17.1', 'ABC13-47656900H15.1', 'ABC7-481722F1.2'\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "Re-querying with the following entries: \n",
      "[]\n",
      "Finished.\n",
      "Remapping complete! 20511/21595 genes remapped.\n",
      "querying 26001-27000...querying 22001-23000...Finished.\n",
      "Remapping complete! 26026/26359 genes remapped.\n",
      "done.done.\n",
      "\n",
      "done.\n",
      "done.\n",
      "done.done.done.\n",
      "\n",
      "\n",
      "done.\n",
      "done.\n",
      "done.done.\n",
      "\n",
      "done.\n",
      "Finished.Finished.\n",
      "\n",
      "1 input query terms found dup hits:querying 20001-21000...\n",
      "\t[('ARMCX5-GPRASP2', 2)]\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "4473 input query terms found dup hits:\n",
      "\t[('TTLL10', 2), ('DVL1', 2), ('MRPL20', 4), ('SSU72', 10), ('PRKCZ', 2), ('SKI', 5), ('TNFRSF14-AS1'\n",
      "2405 input query terms found no hit:\n",
      "\t['AL627309.5', 'AL645608.7', 'AL390719.2', 'AL391244.2', 'AL645728.1', 'AL691432.4', 'AL691432.2', '\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "Remapping complete! 28510/28863 genes remapped.\n",
      "The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "Re-querying with the following entries: \n",
      "['CYB561D2-1', 'HSPA14-1', 'TBCE-1']\n",
      "querying 1-3...Finished.\n",
      "querying 19001-20000...querying 19001-20000...querying 19001-20000...querying 25001-25034...4166 input query terms found dup hits:\n",
      "\t[('TTLL10', 2), ('DVL1', 2), ('MRPL20', 4), ('SSU72', 10), ('SLC35E2', 2), ('PRKCZ', 2), ('SKI', 5),\n",
      "3218 input query terms found no hit:\n",
      "\t['RP11-34P13.7', 'FO538757.2', 'AP006222.2', 'RP4-669L17.10', 'RP5-857K21.4', 'RP11-206L10.9', 'RP11\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "Re-querying with the following entries: \n",
      "[]\n",
      "Finished.\n",
      "Remapping complete! 18823/19681 genes remapped.\n",
      "querying 20001-21000...done.done.\n",
      "\n",
      "querying 32001-33000...querying 23001-24000...done.\n",
      "querying 17001-18000...done.\n",
      "querying 27001-28000...Finished.\n",
      "4877 input query terms found dup hits:\n",
      "\t[('TTLL10', 2), ('DVL1', 2), ('MRPL20', 4), ('SSU72', 10), ('SLC35E2', 2), ('PRKCZ', 2), ('SKI', 5),\n",
      "5906 input query terms found no hit:\n",
      "\t['RP11-34P13.7', 'RP11-34P13.8', 'FO538757.3', 'FO538757.2', 'AP006222.2', 'RP4-669L17.2', 'RP4-669L\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "done.\n",
      "Finished.\n",
      "The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "Re-querying with the following entries: \n",
      "['ATRIP-1', 'C1QTNF9B-AS1-1', 'CCDC7-1', 'CFAP99-1', 'CYB561D2-1', 'ECSCR-1', 'IGHV2-70-1', 'LINC01422-1', 'LINC01481-1', 'PGM5-AS1-1', 'RAET1E-AS1-1', 'SERPINA3-1', 'TIMM10B-1', 'TMEM256-PLSCR3-1']\n",
      "querying 1-14...Remapping complete! 18922/19201 genes remapped.\n",
      "querying 21001-22000...querying 22001-23000...done.\n",
      "done.\n",
      "done.\n",
      "done.done.done.\n",
      "done.\n",
      "\n",
      "\n",
      "done.\n",
      "done.\n",
      "Finished.\n",
      "4 input query terms found dup hits:\n",
      "\t[('ATRIP', 2), ('IGHV2-70', 3), ('LINC01422', 2), ('TMEM256-PLSCR3', 3)]\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "Remapping complete! 24006/25034 genes remapped.\n",
      "querying 20001-21000...querying 23001-24000...done.\n",
      "done.done.\n",
      "\n",
      "querying 24001-25000...querying 21001-22000...querying 20001-21000...querying 20001-21000...querying 20001-21000...querying 21001-22000...done.\n",
      "done.\n",
      "querying 18001-19000...querying 33001-34000...querying 28001-29000...querying 22001-23000...done.\n",
      "querying 23001-24000...done.\n",
      "querying 24001-25000...done.\n",
      "querying 22001-23000...done.\n",
      "done.\n",
      "querying 21001-22000...querying 25001-26000...querying 19001-20000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 21001-22000...done.\n",
      "done.\n",
      "querying 25001-26000...querying 24001-25000...querying 29001-30000...querying 21001-22000...querying 34001-35000...querying 22001-23000...querying 23001-24000...done.\n",
      "done.\n",
      "querying 22001-23000...querying 20001-20926...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 25001-26000...done.\n",
      "done.\n",
      "querying 23001-24000...done.\n",
      "done.\n",
      "querying 30001-31000...done.\n",
      "querying 24001-25000...querying 26001-27000...done.\n",
      "done.\n",
      "querying 22001-23000...querying 26001-26833...done.\n",
      "querying 22001-23000...Finished.\n",
      "4975 input query terms found dup hits:\n",
      "\t[('TTLL10', 2), ('DVL1', 2), ('MRPL20', 4), ('SSU72', 10), ('MMP23A', 3), ('SLC35E2', 2), ('PRKCZ', \n",
      "565 input query terms found no hit:\n",
      "\t['MIR6723', 'LOC100133331', 'LOC100130417', 'LOC148413', 'LOC102724312', 'LOC115110', 'LOC102724450'\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "Re-querying with the following entries: \n",
      "['MIR1254-1']\n",
      "querying 1-1...querying 23001-24000...querying 35001-36000...done.\n",
      "done.\n",
      "done.\n",
      "querying 21001-22000...done.\n",
      "querying 26001-27000...querying 23001-24000...done.\n",
      "done.\n",
      "Finished.\n",
      "Remapping complete! 19815/20926 genes remapped.\n",
      "querying 31001-32000...querying 24001-25000...querying 27001-27613...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 23001-24000...Finished.\n",
      "querying 25001-26000...5462 input query terms found dup hits:\n",
      "\t[('TTLL10', 2), ('LINC01786', 2), ('DVL1', 2), ('MRPL20', 4), ('LINC01770', 2), ('SSU72', 10), ('PRK\n",
      "6643 input query terms found no hit:\n",
      "\t['AL627309.1', 'AL627309.5', 'AP006222.2', 'AL732372.1', 'AL645608.2', 'AL645608.4', 'AL645608.7', 'querying 23001-24000...\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. querying 22001-23000...\n",
      "Re-querying with the following entries: \n",
      "['CYB561D2-1', 'GGT1-1', 'HSPA14-1', 'LINC01238-1', 'TBCE-1']\n",
      "querying 1-5...done.\n",
      "querying 24001-25000...querying 36001-37000...done.\n",
      "querying 27001-28000...done.\n",
      "done.\n",
      "Finished.\n",
      "5138 input query terms found dup hits:\n",
      "\t[('A1BG', 2), ('A2M', 5), ('A2ML1', 5), ('AADACL2', 4), ('AAK1', 2), ('AARSD1', 4), ('ABCA9', 2), ('\n",
      "7408 input query terms found no hit:\n",
      "\t['AB015752.3', 'AB019440.50', 'ABC12-49244600F4.4', 'ABC7-42404400C24.1', 'AC000032.2', 'AC000036.4'\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "Re-querying with the following entries: \n",
      "[]\n",
      "Finished.\n",
      "Remapping complete! 26549/27613 genes remapped.\n",
      "Finished.\n",
      "querying 32001-33000...Remapping complete! 26492/26833 genes remapped.\n",
      "done.done.done.\n",
      "\n",
      "\n",
      "done.\n",
      "done.done.\n",
      "done.\n",
      "\n",
      "done.\n",
      "querying 24001-25000...querying 26001-27000...querying 25001-26000...querying 24001-25000...querying 24001-25000...querying 23001-24000...querying 25001-26000...querying 37001-38000...done.done.\n",
      "\n",
      "querying 33001-34000...querying 28001-29000...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 25001-26000...done.\n",
      "done.\n",
      "querying 27001-28000...querying 26001-26636...querying 38001-39000...querying 24001-25000...querying 26001-26993...done.\n",
      "querying 25001-26000...querying 25001-26000...querying 29001-30000...done.\n",
      "done.\n",
      "done.\n",
      "querying 28001-29000...done.\n",
      "querying 34001-35000...querying 26001-27000...done.\n",
      "done.\n",
      "querying 26001-27000...done.\n",
      "done.\n",
      "done.\n",
      "querying 25001-25740...querying 26001-27000...querying 30001-31000...querying 39001-40000...done.\n",
      "done.\n",
      "Finished.\n",
      "5491 input query terms found dup hits:\n",
      "\t[('A1BG', 2), ('A2M', 5), ('A2ML1', 5), ('A2ML1-AS1', 2), ('AADACL2', 4), ('AADACL2-AS1', 3), ('AAK1\n",
      "5971 input query terms found no hit:\n",
      "\t['AC000032.1', 'AC000036.1', 'AC000061.1', 'AC000068.1', 'AC000068.2', 'AC000068.3', 'AC000403.1', '\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "Re-querying with the following entries: \n",
      "[]\n",
      "Finished.\n",
      "Remapping complete! 26171/26636 genes remapped.\n",
      "done.\n",
      "querying 29001-30000...querying 35001-36000...done.\n",
      "done.\n",
      "done.\n",
      "Finished.\n",
      "5460 input query terms found dup hits:\n",
      "\t[('TTLL10', 2), ('LINC01786', 2), ('DVL1', 2), ('MRPL20', 4), ('LINC01770', 2), ('SSU72', 10), ('PRK\n",
      "6796 input query terms found no hit:\n",
      "\t['AL627309.1', 'AL627309.5', 'AP006222.2', 'AL669831.2', 'AL645608.6', 'AL645608.2', 'AL645608.4', '\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "done.\n",
      "done.\n",
      "The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "Re-querying with the following entries: \n",
      "['CYB561D2-1', 'GGT1-1', 'HSPA14-1', 'LINC01238-1', 'TBCE-1']\n",
      "querying 1-5...done.\n",
      "done.\n",
      "querying 27001-28000...querying 27001-28000...Finished.\n",
      "4925 input query terms found dup hits:\n",
      "\t[('TTLL10', 2), ('DVL1', 2), ('MRPL20', 4), ('SSU72', 10), ('SLC35E2', 2), ('PRKCZ', 2), ('SKI', 5),\n",
      "6359 input query terms found no hit:\n",
      "\t['RP11-34P13.7', 'RP11-34P13.8', 'FO538757.3', 'FO538757.2', 'AP006222.2', 'RP4-669L17.10', 'RP5-857\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "querying 31001-32000...The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "Re-querying with the following entries: \n",
      "['ATRIP-1', 'BLOC1S5-1', 'C1QTNF9B-AS1-1', 'CCDC7-1', 'CFAP99-1', 'CYB561D2-1', 'ECSCR-1', 'IGHV2-70-1', 'LINC00864-1', 'LINC01297-1', 'LINC01422-1', 'LINC01481-1', 'PGM5-AS1-1', 'RAET1E-AS1-1', 'SERPINA3-1', 'TIMM10B-1', 'TMEM256-PLSCR3-1']\n",
      "querying 1-17...querying 27001-28000...querying 40001-41000...done.\n",
      "Finished.\n",
      "Remapping complete! 26654/26993 genes remapped.\n",
      "done.\n",
      "done.\n",
      "querying 36001-37000...querying 30001-31000...Finished.\n",
      "6 input query terms found dup hits:\n",
      "\t[('ATRIP', 2), ('BLOC1S5', 5), ('IGHV2-70', 3), ('LINC01297', 3), ('LINC01422', 2), ('TMEM256-PLSCR3\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "Remapping complete! 24707/25740 genes remapped.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 28001-29000...querying 28001-29000...done.done.\n",
      "\n",
      "done.\n",
      "done.\n",
      "querying 32001-33000...querying 41001-42000...querying 28001-29000...querying 37001-38000...querying 31001-31399...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 29001-29146...done.\n",
      "done.\n",
      "done.\n",
      "done.\n",
      "querying 29001-30000...querying 33001-34000...Finished.\n",
      "7237 input query terms found dup hits:\n",
      "\t[('A1BG', 2), ('A2M', 5), ('AAK1', 2), ('AARSD1', 4), ('ABCA9', 2), ('ABCB10', 10), ('ABCC2', 2), ('\n",
      "9253 input query terms found no hit:\n",
      "\t['AC000068.10', 'AC000068.5', 'AC000068.9', 'AC002044.4', 'AC002064.5', 'AC002066.1', 'AC002116.7', \n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "Re-querying with the following entries: \n",
      "['ERCC-00019', 'ERCC-00024', 'ERCC-00028', 'ERCC-00034', 'ERCC-00083', 'ERCC-00086', 'ERCC-00095', 'ERCC-00096', 'ERCC-00104', 'ERCC-00137', 'ERCC-00138', 'ERCC-00150', 'ERCC-00165', 'ERCC-00168', 'ERCC-00171', 'MIR1254-1', 'hsa-mir-150', 'snoU2-30']\n",
      "querying 1-18...querying 38001-39000...querying 29001-30000...querying 42001-42665...Finished.\n",
      "7240 input query terms found dup hits:\n",
      "\t[('A1BG', 2), ('A2M', 5), ('A2ML1', 5), ('AADACL2', 4), ('AADACL2-AS1', 3), ('AAK1', 2), ('AARSD1', \n",
      "7101 input query terms found no hit:\n",
      "\t['ABBA01000935.2', 'AC000032.1', 'AC000061.1', 'AC000067.1', 'AC000068.1', 'AC000068.2', 'AC000077.1\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "done.\n",
      "The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "Re-querying with the following entries: \n",
      "['MIR1254-1']\n",
      "querying 1-1...done.\n",
      "done.\n",
      "Finished.\n",
      "2 input query terms found dup hits:\n",
      "\t[('ERCC', 135), ('hsa-mir', 2)]\n",
      "1 input query terms found no hit:\n",
      "\t['snoU2']\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "Remapping complete! 30017/31399 genes remapped.\n",
      "done.\n",
      "Finished.\n",
      "Remapping complete! 28512/29146 genes remapped.\n",
      "querying 30001-31000...done.\n",
      "done.\n",
      "querying 39001-40000...done.\n",
      "querying 34001-35000...Finished.\n",
      "11879 input query terms found dup hits:\n",
      "\t[('SCYL3', 2), ('SEMA3F', 2), ('CFTR', 10), ('CYP51A1', 2), ('RAD52', 4), ('MYH16', 2), ('LAP3', 2),\n",
      "13832 input query terms found no hit:\n",
      "\t['ENSG00000036549', 'ENSG00000100101', 'ENSG00000111780', 'ENSG00000111788', 'ENSG00000115934', 'ENS\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "done.\n",
      "querying 30001-31000...The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "Re-querying with the following entries: \n",
      "['MIR3687-2']\n",
      "querying 1-1...done.\n",
      "done.\n",
      "querying 31001-32000...done.\n",
      "Finished.\n",
      "1 input query terms found no hit:\n",
      "querying 40001-41000...\t['MIR3687']\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "Remapping complete! 42057/42665 genes remapped.\n",
      "querying 31001-32000...done.\n",
      "done.\n",
      "querying 35001-36000...querying 32001-33000...done.\n",
      "querying 41001-42000...done.done.\n",
      "\n",
      "done.\n",
      "querying 33001-34000...querying 32001-33000...querying 36001-37000...done.\n",
      "querying 42001-43000...done.\n",
      "done.\n",
      "done.\n",
      "querying 37001-38000...querying 33001-34000...querying 34001-35000...done.\n",
      "querying 43001-44000...done.\n",
      "done.\n",
      "done.\n",
      "querying 38001-39000...done.\n",
      "querying 34001-35000...querying 35001-36000...querying 44001-45000...done.\n",
      "querying 39001-40000...done.\n",
      "done.\n",
      "querying 45001-46000...done.\n",
      "done.\n",
      "querying 35001-36000...querying 36001-36900...querying 40001-41000...done.\n",
      "done.\n",
      "querying 46001-47000...done.\n",
      "querying 36001-37000...done.\n",
      "Finished.\n",
      "7492 input query terms found dup hits:\n",
      "\t[('5S_rRNA', 10), ('5_8S_rRNA', 9), ('7SK', 10), ('A1BG', 2), ('A2M', 5), ('A2ML1', 5), ('A2ML1-AS1'\n",
      "11979 input query terms found no hit:\n",
      "\t['AB015752.1', 'AC000032.1', 'AC000035.1', 'AC000036.1', 'AC000058.1', 'AC000061.1', 'AC000067.1', '\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "Re-querying with the following entries: \n",
      "['Clostridiales-1', 'MIR1254-1', 'hsa-mir-3119-1', 'hsa-mir-3158-1', 'hsa-mir-548d-2', 'snoMBII-202', 'snoU2-30']\n",
      "querying 1-7...querying 41001-42000...done.\n",
      "done.\n",
      "done.\n",
      "Finished.\n",
      "6 input query terms found no hit:\n",
      "\t['Clostridiales', 'hsa-mir-3119', 'hsa-mir-3158', 'hsa-mir-548d', 'snoMBII', 'snoU2']\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "Remapping complete! 36172/36900 genes remapped.\n",
      "querying 47001-48000...querying 37001-38000...done.\n",
      "done.\n",
      "querying 42001-43000...querying 48001-48353...done.\n",
      "done.\n",
      "querying 38001-39000...done.\n",
      "Finished.\n",
      "13805 input query terms found dup hits:\n",
      "\t[('PARTICL', 2), ('H3C2', 10), ('CCDC15-DT', 2), ('LINC02717', 2), ('FAM245B', 2), ('LINC02838', 2),\n",
      "17125 input query terms found no hit:\n",
      "\t['AL645933.5', 'AC010184.1', 'AC023296.1', 'AL117334.2', 'AL357079.3', 'AL442063.1', 'AL360015.1', '\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "Re-querying with the following entries: \n",
      "[]\n",
      "Finished.\n",
      "Remapping complete! 48095/48353 genes remapped.\n",
      "querying 43001-44000...done.\n",
      "querying 39001-40000...done.\n",
      "querying 44001-45000...done.\n",
      "querying 40001-41000...done.\n",
      "querying 45001-46000...done.\n",
      "done.\n",
      "querying 46001-46813...querying 41001-42000...done.\n",
      "done.\n",
      "Finished.\n",
      "13805 input query terms found dup hits:\n",
      "\t[('A1BG', 2), ('A2M', 5), ('A2ML1', 5), ('A2ML1-AS2', 2), ('AADACL2', 4), ('AADACL2-AS1', 3), ('AAK1\n",
      "15107 input query terms found no hit:\n",
      "\t['AB015752.1', 'AC000032.1', 'AC000036.1', 'AC000058.1', 'AC000061.1', 'AC000067.1', 'AC000068.1', '\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "Re-querying with the following entries: \n",
      "['MIR1254-1', 'MIR7641-1']\n",
      "querying 1-2...querying 42001-43000...done.\n",
      "Finished.\n",
      "1 input query terms found no hit:\n",
      "\t['MIR7641']\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "Remapping complete! 46173/46813 genes remapped.\n",
      "done.\n",
      "querying 43001-44000...done.\n",
      "querying 44001-45000...done.\n",
      "querying 45001-46000...done.\n",
      "querying 46001-47000...done.\n",
      "querying 47001-48000...done.\n",
      "querying 48001-49000...done.\n",
      "querying 49001-50000...done.\n",
      "querying 50001-51000...done.\n",
      "querying 51001-52000...done.\n",
      "querying 52001-53000...done.\n",
      "querying 53001-54000...done.\n",
      "querying 54001-55000...done.\n",
      "querying 55001-55556...done.\n",
      "Finished.\n",
      "16414 input query terms found dup hits:\n",
      "\t[('PARTICL', 2), ('H3C2', 10), ('CCDC15-DT', 2), ('LINC02717', 2), ('FAM245B', 2), ('LINC02838', 2),\n",
      "21042 input query terms found no hit:\n",
      "\t['AL645933.5', 'AC010184.1', 'AC023296.1', 'AL117334.2', 'AC107294.4', 'AL357079.3', 'AL442063.1', '\n",
      "Pass \"returnall=True\" to return complete lists of duplicate or missing query terms.\n",
      "The following genes were not found and a '-\\d+$' pattern was detected. \n",
      "They could potentially be affected by `.var_names_make_unique()`. \n",
      "Re-querying with the following entries: \n",
      "[]\n",
      "Finished.\n",
      "Remapping complete! 55286/55556 genes remapped.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "datasets_remapped = process_map(remap_gene_symbols, datasets.values(), max_workers=32)\n",
    "for dataset_id, dataset in zip(datasets.keys(), datasets_remapped):\n",
    "    datasets[dataset_id] = dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### aggregate duplicate gene symbols\n",
    "\n",
    "TODO: take care of duplicate genes that were made unique with \"var_names_make_unique\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adams_Kaminski_2020_COPD\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 28/28 [00:00<00:00, 273.52it/s]\n",
      "/home/sturm/.conda/envs/pircher-sc-integrate2/lib/python3.8/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chen_Zhang_2020_NSCLC\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 19/19 [00:00<00:00, 364.23it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Goveia_Carmeliet_2020_NSCLC\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 22/22 [00:00<00:00, 501.60it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Guo_Zhang_2018_NSCLC\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 20/20 [00:00<00:00, 275.15it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Habermann_Kropski_2020_pulmonary-fibrosis\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 49/49 [00:00<00:00, 430.66it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "He_Fan_2021_LUAD\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 47/47 [00:00<00:00, 416.33it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Kim_Lee_2020_LUAD\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 16/16 [00:00<00:00, 419.83it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Lambrechts_2018_LUAD_6149v1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 9/9 [00:00<00:00, 548.78it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Lambrechts_2018_LUAD_6149v2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 16/16 [00:00<00:00, 371.11it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Lambrechts_2018_LUAD_6653\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 14/14 [00:00<00:00, 443.97it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Laughney_Massague_2020_NSCLC\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 30/30 [00:00<00:00, 535.45it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Madissoon_Meyer_2020_pulmonary-fibrosis\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 53/53 [00:00<00:00, 465.87it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Maier_Merad_2020_NSCLC\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 27/27 [00:00<00:00, 592.00it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Maynard_Bivona_2020_NSCLC\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 20/20 [00:00<00:00, 244.16it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mayr_Schiller_2020_pulmonary-fibrosis\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 785/785 [00:01<00:00, 419.48it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reyfman_Misharin_2018_pulmonary-fibrosis\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 58/58 [00:00<00:00, 478.31it/s]\n",
      "100%|██████████| 44/44 [00:00<00:00, 582.54it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Travaglini_Krasnow_2020_Lung_10x\n",
      "Travaglini_Krasnow_2020_Lung_SS2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "100%|██████████| 35/35 [00:00<00:00, 293.14it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UKIM-V\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 15/15 [00:00<00:00, 455.59it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Vieira_Teichmann_2019_asthma\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 31/31 [00:00<00:00, 680.01it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wu_Zhou_2021_NSCLC\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 21/21 [00:00<00:00, 404.72it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Zilionis_Klein_2019_NSCLC\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 32/32 [00:00<00:00, 372.62it/s]\n"
     ]
    }
   ],
   "source": [
    "for dataset_id, dataset in datasets.items():\n",
    "    print(dataset_id)\n",
    "    datasets[dataset_id] = aggregate_duplicate_gene_symbols(dataset)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## add cell type annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "for dataset in datasets.values():\n",
    "    dataset.obs[\"cell_type\"] = \"unknown\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_all = pd.concat([x.obs for x in datasets.values()]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_all = obs_all.loc[\n",
    "    :,\n",
    "    MANDATORY_COLS\n",
    "    + [\n",
    "        \"accession\",\n",
    "        \"sampleType\",\n",
    "        \"platform\",\n",
    "        \"age\",\n",
    "        \"tobacco\",\n",
    "        \"ethnicity\",\n",
    "        \"processing_site\",\n",
    "        \"Tissue origins\",\n",
    "        \"histology\",\n",
    "        \"smoking\",\n",
    "        \"pathology\",\n",
    "        \"EGFR\",\n",
    "        \"tumor_stage\",\n",
    "        \"geo_accession\",\n",
    "        \"tissue_orig\",\n",
    "        \"replicate\",\n",
    "        \"race\",\n",
    "        \"smoking_status\",\n",
    "        \"driver_gene\",\n",
    "        \"driver_mutation\",\n",
    "        \"secondary_mutation\",\n",
    "        \"Notes\",\n",
    "        \"stage_at_diagnosis\",\n",
    "        \"pathlogy_review\",\n",
    "        \"biopsy_date\",\n",
    "        \"sort_date\",\n",
    "        \"biopsy_type\",\n",
    "        \"biopsy_time_status\",\n",
    "        \"early_treatment_status\",\n",
    "        \"best_response_status\",\n",
    "        \"biopsy_timing\",\n",
    "        \"analysis\",\n",
    "        \"treatment_history\",\n",
    "        \"treatment_history_detail\",\n",
    "        \"line_of_therapy\",\n",
    "        \"treatment_type\",\n",
    "        \"treatment\",\n",
    "        \"percent_PFS_ref_values\",\n",
    "        \"percent.PFS.reference.values\",\n",
    "        \"infections\",\n",
    "        \"early_bx_day\",\n",
    "        \"treatment_start_date\",\n",
    "        \"pfs_over_under\",\n",
    "        \"pfs_day\",\n",
    "        \"pfs_month\",\n",
    "        \"date_of_death\",\n",
    "        \"stageIII.IV_ca_dx_date\",\n",
    "        \"ca_dx_OS\",\n",
    "        \"region\",\n",
    "        \"location\",\n",
    "        \"label\",\n",
    "        \"tumor_id\",\n",
    "        \"tumor_type\",\n",
    "        \"GEO_Sample\",\n",
    "        \"biopsy_segment\",\n",
    "        \"gsm\",\n",
    "        \"characteristics_ch1.7.treatment received prior to surgery (1= treated; 0=untreated)\",\n",
    "    ],\n",
    "].drop_duplicates().set_index(\"sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_all.to_excel(\"../../data/50_integrate_scrnaseq_data/51_merge_all/obs_all.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_all = merge_datasets(datasets.values(), symbol_in_n_datasets=17)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1124947, 17833)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_all.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>patient</th>\n",
       "      <th>tissue</th>\n",
       "      <th>origin</th>\n",
       "      <th>condition</th>\n",
       "      <th>dataset</th>\n",
       "      <th>sex</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Adams_Kaminski_2020_COPD_001C</td>\n",
       "      <td>Adams_Kaminski_2020_COPD_001C</td>\n",
       "      <td>lung</td>\n",
       "      <td>normal</td>\n",
       "      <td>healthy_control</td>\n",
       "      <td>Adams_Kaminski_2020_COPD</td>\n",
       "      <td>nan</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Adams_Kaminski_2020_COPD_Adams_Kaminski_2020_C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Adams_Kaminski_2020_COPD_002C</td>\n",
       "      <td>Adams_Kaminski_2020_COPD_002C</td>\n",
       "      <td>lung</td>\n",
       "      <td>normal</td>\n",
       "      <td>healthy_control</td>\n",
       "      <td>Adams_Kaminski_2020_COPD</td>\n",
       "      <td>nan</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Adams_Kaminski_2020_COPD_Adams_Kaminski_2020_C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Adams_Kaminski_2020_COPD_003C</td>\n",
       "      <td>Adams_Kaminski_2020_COPD_003C</td>\n",
       "      <td>lung</td>\n",
       "      <td>normal</td>\n",
       "      <td>healthy_control</td>\n",
       "      <td>Adams_Kaminski_2020_COPD</td>\n",
       "      <td>nan</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Adams_Kaminski_2020_COPD_Adams_Kaminski_2020_C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Adams_Kaminski_2020_COPD_022C-a</td>\n",
       "      <td>Adams_Kaminski_2020_COPD_222C</td>\n",
       "      <td>lung</td>\n",
       "      <td>normal</td>\n",
       "      <td>healthy_control</td>\n",
       "      <td>Adams_Kaminski_2020_COPD</td>\n",
       "      <td>nan</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Adams_Kaminski_2020_COPD_Adams_Kaminski_2020_C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Adams_Kaminski_2020_COPD_022C-b</td>\n",
       "      <td>Adams_Kaminski_2020_COPD_222C</td>\n",
       "      <td>lung</td>\n",
       "      <td>normal</td>\n",
       "      <td>healthy_control</td>\n",
       "      <td>Adams_Kaminski_2020_COPD</td>\n",
       "      <td>nan</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Adams_Kaminski_2020_COPD_Adams_Kaminski_2020_C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>500</th>\n",
       "      <td>Zilionis_Klein_2019_NSCLC_p2t1</td>\n",
       "      <td>Zilionis_Klein_2019_NSCLC_patient_2</td>\n",
       "      <td>lung</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>LSCC</td>\n",
       "      <td>Zilionis_Klein_2019_NSCLC</td>\n",
       "      <td>female</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Zilionis_Klein_2019_NSCLC_Zilionis_Klein_2019_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>501</th>\n",
       "      <td>Zilionis_Klein_2019_NSCLC_p4t1</td>\n",
       "      <td>Zilionis_Klein_2019_NSCLC_patient_4</td>\n",
       "      <td>lung</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Zilionis_Klein_2019_NSCLC</td>\n",
       "      <td>male</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Zilionis_Klein_2019_NSCLC_Zilionis_Klein_2019_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>502</th>\n",
       "      <td>Zilionis_Klein_2019_NSCLC_p1t3</td>\n",
       "      <td>Zilionis_Klein_2019_NSCLC_patient_1</td>\n",
       "      <td>lung</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>LSCC</td>\n",
       "      <td>Zilionis_Klein_2019_NSCLC</td>\n",
       "      <td>male</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Zilionis_Klein_2019_NSCLC_Zilionis_Klein_2019_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>503</th>\n",
       "      <td>Zilionis_Klein_2019_NSCLC_p5t1</td>\n",
       "      <td>Zilionis_Klein_2019_NSCLC_patient_5</td>\n",
       "      <td>lung</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Zilionis_Klein_2019_NSCLC</td>\n",
       "      <td>female</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Zilionis_Klein_2019_NSCLC_Zilionis_Klein_2019_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>504</th>\n",
       "      <td>Zilionis_Klein_2019_NSCLC_p3t3</td>\n",
       "      <td>Zilionis_Klein_2019_NSCLC_patient_3</td>\n",
       "      <td>lung</td>\n",
       "      <td>tumor_primary</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>Zilionis_Klein_2019_NSCLC</td>\n",
       "      <td>female</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Zilionis_Klein_2019_NSCLC_Zilionis_Klein_2019_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>505 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              sample                              patient  \\\n",
       "0      Adams_Kaminski_2020_COPD_001C        Adams_Kaminski_2020_COPD_001C   \n",
       "1      Adams_Kaminski_2020_COPD_002C        Adams_Kaminski_2020_COPD_002C   \n",
       "2      Adams_Kaminski_2020_COPD_003C        Adams_Kaminski_2020_COPD_003C   \n",
       "3    Adams_Kaminski_2020_COPD_022C-a        Adams_Kaminski_2020_COPD_222C   \n",
       "4    Adams_Kaminski_2020_COPD_022C-b        Adams_Kaminski_2020_COPD_222C   \n",
       "..                               ...                                  ...   \n",
       "500   Zilionis_Klein_2019_NSCLC_p2t1  Zilionis_Klein_2019_NSCLC_patient_2   \n",
       "501   Zilionis_Klein_2019_NSCLC_p4t1  Zilionis_Klein_2019_NSCLC_patient_4   \n",
       "502   Zilionis_Klein_2019_NSCLC_p1t3  Zilionis_Klein_2019_NSCLC_patient_1   \n",
       "503   Zilionis_Klein_2019_NSCLC_p5t1  Zilionis_Klein_2019_NSCLC_patient_5   \n",
       "504   Zilionis_Klein_2019_NSCLC_p3t3  Zilionis_Klein_2019_NSCLC_patient_3   \n",
       "\n",
       "    tissue         origin        condition                    dataset     sex  \\\n",
       "0     lung         normal  healthy_control   Adams_Kaminski_2020_COPD     nan   \n",
       "1     lung         normal  healthy_control   Adams_Kaminski_2020_COPD     nan   \n",
       "2     lung         normal  healthy_control   Adams_Kaminski_2020_COPD     nan   \n",
       "3     lung         normal  healthy_control   Adams_Kaminski_2020_COPD     nan   \n",
       "4     lung         normal  healthy_control   Adams_Kaminski_2020_COPD     nan   \n",
       "..     ...            ...              ...                        ...     ...   \n",
       "500   lung  tumor_primary             LSCC  Zilionis_Klein_2019_NSCLC  female   \n",
       "501   lung  tumor_primary             LUAD  Zilionis_Klein_2019_NSCLC    male   \n",
       "502   lung  tumor_primary             LSCC  Zilionis_Klein_2019_NSCLC    male   \n",
       "503   lung  tumor_primary             LUAD  Zilionis_Klein_2019_NSCLC  female   \n",
       "504   lung  tumor_primary             LUAD  Zilionis_Klein_2019_NSCLC  female   \n",
       "\n",
       "    cell_type                                              batch  \n",
       "0     unknown  Adams_Kaminski_2020_COPD_Adams_Kaminski_2020_C...  \n",
       "1     unknown  Adams_Kaminski_2020_COPD_Adams_Kaminski_2020_C...  \n",
       "2     unknown  Adams_Kaminski_2020_COPD_Adams_Kaminski_2020_C...  \n",
       "3     unknown  Adams_Kaminski_2020_COPD_Adams_Kaminski_2020_C...  \n",
       "4     unknown  Adams_Kaminski_2020_COPD_Adams_Kaminski_2020_C...  \n",
       "..        ...                                                ...  \n",
       "500   unknown  Zilionis_Klein_2019_NSCLC_Zilionis_Klein_2019_...  \n",
       "501   unknown  Zilionis_Klein_2019_NSCLC_Zilionis_Klein_2019_...  \n",
       "502   unknown  Zilionis_Klein_2019_NSCLC_Zilionis_Klein_2019_...  \n",
       "503   unknown  Zilionis_Klein_2019_NSCLC_Zilionis_Klein_2019_...  \n",
       "504   unknown  Zilionis_Klein_2019_NSCLC_Zilionis_Klein_2019_...  \n",
       "\n",
       "[505 rows x 9 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_all.obs.drop_duplicates().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'sample' as categorical\n",
      "... storing 'patient' as categorical\n",
      "... storing 'tissue' as categorical\n",
      "... storing 'origin' as categorical\n",
      "... storing 'condition' as categorical\n",
      "... storing 'dataset' as categorical\n",
      "... storing 'sex' as categorical\n",
      "... storing 'cell_type' as categorical\n",
      "... storing 'batch' as categorical\n"
     ]
    }
   ],
   "source": [
    "merged_all.write_h5ad(\n",
    "    \"../../data/50_integrate_scrnaseq_data/51_merge_all/merged_all.h5ad\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export for NSCLC heterogeneity\n",
    " * only tumor samples (no controls, no metastases)\n",
    " * all NSCLC subtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# datasets_nsclc_heterogeneity = dict()\n",
    "# for dataset_id, dataset in datasets.items():\n",
    "#     if \"tumor_primary\" in dataset.obs[\"origin\"].values:\n",
    "#         datasets_nsclc_heterogeneity[dataset_id] = dataset[\n",
    "#             dataset.obs[\"origin\"] == \"tumor_primary\", :\n",
    "#         ].copy()\n",
    "# del datasets_nsclc_heterogeneity[\"Pircher_batch1_NSCLC\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merged_nsclc_heterogeneity = merge_datasets(\n",
    "#     datasets_nsclc_heterogeneity.values(), symbol_in_n_datasets=5\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merged_nsclc_heterogeneity.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merged_nsclc_heterogeneity.write_h5ad(\n",
    "#     \"../../data/50_integrate_scrnaseq_data/51_merge_all/merged_nsclc_heterogeneity.h5ad\"\n",
    "# )"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "text_representation": {
    "extension": ".py",
    "format_name": "percent",
    "format_version": "1.3",
    "jupytext_version": "1.11.2"
   }
  },
  "kernelspec": {
   "display_name": "Python [conda env:.conda-pircher-sc-integrate2]",
   "language": "python",
   "name": "conda-env-.conda-pircher-sc-integrate2-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
